<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>nbaR: R access to the Netherlands Biodiversity API • nbaR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="nbaR: R access to the Netherlands Biodiversity API">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nbaR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/nbaR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/sharks.html">Calibrating a molecular phylogeny of the sharks</a>
    </li>
    <li>
      <a href="../articles/tomato.html">The oldest tomato specimen in the world</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/naturalis/nbaR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>nbaR: R access to the Netherlands Biodiversity API</h1>
                        <h4 class="author">Hannes Hettling</h4>
            
            <h4 class="date">Oct 02, 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/naturalis/nbaR/blob/master/vignettes/nbaR.Rmd"><code>vignettes/nbaR.Rmd</code></a></small>
      <div class="hidden name"><code>nbaR.Rmd</code></div>

    </div>

    
    
<div id="background" class="section level1">
<h1 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h1>
<div id="the-netherlands-biodiversity-api" class="section level2">
<h2 class="hasAnchor">
<a href="#the-netherlands-biodiversity-api" class="anchor"></a>The Netherlands Biodiversity API</h2>
<p>The Netherlands Biodiversity API (NBA) facilitates access to the Natural History Collection at Naturalis Biodiversity Center. Next to museum specimen records and metadata, access to taxonomic classification and nomenclature, to geographical information, and to multimedia files is provided. By using the powerful Elasticsearch engine, the NBA facilitates searching for collection- and biodiversity data in near real-time. Furthermore, by incorporating information from taxonomic databases, taxonomic name resolution can be accomplished with the NBA. Persistent Uniform Resource Identifiers (PURLs) ensure that each specimen accessible via the NBA is represented by a citeable unambiguous web reference. Access to our data is provided via a RESTful interface and several clients such as the <a href="http://bioportal.naturalis.nl/">BioPortal</a>, a web application for browsing biodiversity data that is served by the NBA. For more information about the NBA, please see our <a href="http://docs.biodiversitydata.nl/">detailed documentation</a>.</p>
</div>
<div id="r-access" class="section level2">
<h2 class="hasAnchor">
<a href="#r-access" class="anchor"></a>R access</h2>
<p>The R programming language is established as a common tool in scientific research, with growing adoption by researchers in biodiversity research. Hence, to ease the access to the NBA for researchers, we developed this R client.</p>
</div>
</div>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h1>
<div id="main-data-types-and-classes" class="section level2">
<h2 class="hasAnchor">
<a href="#main-data-types-and-classes" class="anchor"></a>Main data types and classes</h2>
<p>The data served by the NBA consists of four main data types:</p>
<ul>
<li>Specimen</li>
<li>Taxon</li>
<li>Multimedia</li>
<li>Geo</li>
</ul>
<p>Additionally, the data type <em>Metadata</em> stores miscellaneous information about NBA settings. Each of the data types is modelled as an <code>R6</code> class and therefore has its own members such as fields and methods. Documentation about a specific class can be retrieved in the standard manner, e.g. <code><a href="../reference/Specimen.html">?Specimen</a></code> Each class of the data model can be instantiated and has a <code>toJSONString</code> and <code>toList</code> method returning the object’s JSON representation and the object’s data as a list datatype, respectively.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># load nbaR</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">library</span>(nbaR) </a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co"># instanciate specimen object</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5">spec &lt;-<span class="st"> </span>Specimen<span class="op">$</span><span class="kw">new</span>()</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># represent data as JSON or list</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">spec<span class="op">$</span><span class="kw">toJSONString</span>()</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">spec<span class="op">$</span><span class="kw">toList</span>()</a></code></pre></div>
</div>
<div id="api-client-classes" class="section level2">
<h2 class="hasAnchor">
<a href="#api-client-classes" class="anchor"></a>API Client classes</h2>
<p>The interaction with the API is accomplished by the API client classes:</p>
<ul>
<li>SpecimenClient</li>
<li>TaxonClient</li>
<li>MultimediaClient</li>
<li>GeoClient</li>
<li>MetadataClient</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># initialize client</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">client &lt;-<span class="st"> </span>SpecimenClient<span class="op">$</span><span class="kw">new</span>()</a></code></pre></div>
<p>The client class is by default initialized to connect to the base URL <a href="http://api.biodiversitydata.nl/v2" class="uri">http://api.biodiversitydata.nl/v2</a>. For testing purposes, this can be set to a different URL, see <code><a href="../reference/SpecimenClient.html">?SpecimenClient</a></code> for details.</p>
</div>
<div id="queries" class="section level2">
<h2 class="hasAnchor">
<a href="#queries" class="anchor"></a>Queries</h2>
<div id="concept" class="section level3">
<h3 class="hasAnchor">
<a href="#concept" class="anchor"></a>Concept</h3>
<p>With the SpecimentClient created above, the <code>query</code> endpoint for specimens can now be reached via the <code>query</code> function.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># retrieve results of empty query.</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">res &lt;-<span class="st"> </span>client<span class="op">$</span><span class="kw">query</span>()</a></code></pre></div>
<p>The <code>query</code> function then returns an object of class <code>Response</code>, which, in turn, has a field <code>content</code> of class <code>QueryResult</code>. From the <code>QueryResult</code>, the single result items can be accessed as follows:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># get first element of results</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">spec &lt;-<span class="st"> </span>res<span class="op">$</span>content<span class="op">$</span>resultSet[[<span class="dv">1</span>]]<span class="op">$</span>item</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co"># object should be of type Specimen</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw">class</span>(spec)</a></code></pre></div>
<pre><code>## [1] "Specimen" "R6"</code></pre>
</div>
<div id="passing-query-parameters-as-list" class="section level3">
<h3 class="hasAnchor">
<a href="#passing-query-parameters-as-list" class="anchor"></a>Passing query parameters as list</h3>
<p>Query parameters can be specified as a list with named parameters. To query for instance for specimen records that have the type status holotype and a female sex, one can pass a named list as the <code>queryParams</code> parameter to the <code>query</code> function:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co"># specify two query conditions</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">l &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">identifications.typeStatus=</span><span class="st">"holotype"</span>, <span class="dt">sex=</span><span class="st">"female"</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co"># run query</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">res &lt;-<span class="st"> </span>client<span class="op">$</span><span class="kw">query</span>(<span class="dt">queryParams=</span>l)</a></code></pre></div>
<p>Note that by default, both conditions are connected by a logical <code>AND</code>. The <code>queryParams</code> passed as a list thus correspond to basic <a href="http://docs.biodiversitydata.nl/en/latest/quickstart/#human-readable"><em>human readable</em> queries</a>. For more advanced queries, containing different logical operators or nested sub-queries, the user can specify the query in a <code>QuerySpec</code> object.</p>
</div>
</div>
</div>
<div id="advanced-queries" class="section level1">
<h1 class="hasAnchor">
<a href="#advanced-queries" class="anchor"></a>Advanced queries</h1>
<div id="using-the-queryspec-object" class="section level2">
<h2 class="hasAnchor">
<a href="#using-the-queryspec-object" class="anchor"></a>Using the QuerySpec object</h2>
<p>Advanced queries with different operators than <code>AND</code> or nested query conditions can not be accomplished by simply passing the query parameters as a list. Instead, a query is modeled as a <code>QuerySpec</code> object which captures the relationships between multiple query terms. Please also refer to the <a href="http://docs.biodiversitydata.nl/en/latest/advanced-queries/#queryspec">NBA QuerySpec documentation</a> for more information.</p>
<p>A <code>QuerySpec</code> object usually consists of one or more <code>QueryCondition</code> objects, specifying query terms. A <code>QueryCondition</code> object usually contains the fields <code>field</code>, <code>operator</code>, and <code>value</code>(see also <code><a href="../reference/QueryCondition.html">?QueryCondition</a></code>). These fields can be specified in the constructor. If, for example, we want to query for records with a unitID equal to <em>L.4304195</em>, a <code>QueryCondition</code> would look as follows:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co"># specify search in QueryCondition object</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"unitID"</span>, <span class="dt">operator=</span><span class="st">"EQUALS"</span>, <span class="dt">value=</span><span class="st">"L.4304195"</span>)</a></code></pre></div>
<p>Note: The function <code>getFieldInfo</code> on a certain field lists which operators are allowed for that field (e.g. <code>sc$get_field_info()$content$unitID$allowedOperators</code>).</p>
<p>Now, a <code>QuerySpec</code> object can be assembled with the <code>QueryCondition</code>(s) passed as a list:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co"># build QuerySpec using above conditions</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(qc))</a>
<a class="sourceLine" id="cb9-3" data-line-number="3"></a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="co"># do the query</span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5">res &lt;-<span class="st"> </span>client<span class="op">$</span><span class="kw">query</span>(<span class="dt">querySpec=</span>qs)</a></code></pre></div>
<p>Below, we show an example of how to nest multiple query conditions. The query conditions below define to query for specimens of sex ** female* and family <em>Equidae</em> and of rank <em>Species</em>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># specify multiple conditions</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">q1 &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span> <span class="st">"sex"</span>, <span class="dt">operator=</span><span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">                         <span class="dt">value=</span><span class="st">"female"</span>)</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">q2 &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"identifications.defaultClassification.family"</span>,</a>
<a class="sourceLine" id="cb10-5" data-line-number="5">                         <span class="dt">operator=</span><span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb10-6" data-line-number="6">                         <span class="dt">value=</span><span class="st">"Equidae"</span>)</a>
<a class="sourceLine" id="cb10-7" data-line-number="7">q3 &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"identifications.taxonRank"</span>,</a>
<a class="sourceLine" id="cb10-8" data-line-number="8">                         <span class="dt">operator=</span><span class="st">"EQUALS"</span>, <span class="dt">value=</span><span class="st">"species"</span>)</a></code></pre></div>
<p>Extending the constraint to also include specimens of rank <em>Subspecies</em>,<br>
we can combine the latter condition with an additional one using the method <code>or</code>:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># logical conjunction using 'or' method</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2">q3<span class="op">$</span>or &lt;-<span class="st"> </span><span class="kw">list</span>(QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span> <span class="st">"identifications.taxonRank"</span>,</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">                                <span class="dt">operator=</span><span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">                                <span class="dt">value=</span><span class="st">"subspecies"</span>))</a>
<a class="sourceLine" id="cb11-5" data-line-number="5"></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="co"># build QuerySpec from QueryConditions</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7">qs &lt;-QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(q1, q2, q3))</a>
<a class="sourceLine" id="cb11-8" data-line-number="8"></a>
<a class="sourceLine" id="cb11-9" data-line-number="9"><span class="co"># call API</span></a>
<a class="sourceLine" id="cb11-10" data-line-number="10">res &lt;-client<span class="op">$</span><span class="kw">query</span>(<span class="dt">querySpec=</span>qs)</a></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># get 1000 specimen items from plant order 'Gentianales'</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span> <span class="st">"identifications.defaultClassification.order"</span>,</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">                         <span class="dt">operator=</span><span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb12-4" data-line-number="4">                         <span class="dt">value=</span><span class="st">"Gentianales"</span>)</a>
<a class="sourceLine" id="cb12-5" data-line-number="5">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(qc), <span class="dt">size=</span><span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb12-6" data-line-number="6">res &lt;-<span class="st"> </span>client<span class="op">$</span><span class="kw">query</span>(<span class="dt">querySpec=</span>qs)</a>
<a class="sourceLine" id="cb12-7" data-line-number="7"></a>
<a class="sourceLine" id="cb12-8" data-line-number="8"><span class="co"># extract orders from first identification of each specimen</span></a>
<a class="sourceLine" id="cb12-9" data-line-number="9">families &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">sapply</span>(res<span class="op">$</span>content<span class="op">$</span>resultSet,</a>
<a class="sourceLine" id="cb12-10" data-line-number="10">                          <span class="cf">function</span>(x)x<span class="op">$</span>item<span class="op">$</span>identifications[[<span class="dv">1</span>]]<span class="op">$</span>defaultClassification<span class="op">$</span>family))</a>
<a class="sourceLine" id="cb12-11" data-line-number="11"><span class="co"># plot frequencies</span></a>
<a class="sourceLine" id="cb12-12" data-line-number="12"><span class="kw">barplot</span>(<span class="kw">table</span>(families), <span class="dt">horiz=</span>T,</a>
<a class="sourceLine" id="cb12-13" data-line-number="13">        <span class="dt">col=</span><span class="kw">rainbow</span>(<span class="kw">length</span>(<span class="kw">unique</span>(families))),</a>
<a class="sourceLine" id="cb12-14" data-line-number="14">        <span class="dt">las=</span><span class="dv">2</span>, <span class="dt">xlab=</span><span class="st">"frequency"</span>, <span class="dt">cex.names=</span><span class="fl">0.5</span>)</a></code></pre></div>
<p><img src="nbaR_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div id="services-examples" class="section level1">
<h1 class="hasAnchor">
<a href="#services-examples" class="anchor"></a>Services examples</h1>
<div id="specimen-occurrence-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#specimen-occurrence-services" class="anchor"></a>Specimen occurrence services</h2>
<p>Specimen records constitute the core of data served by the NBA. Museum specimens can represent a whole variety of different objects such as plants, animals or single parts thereof, DNA samples, fossils, rocks or meteorites. For detailed information of the data model, please refer to the <a href="http://docs.biodiversitydata.nl/en/latest/doc-spec-services/specimen/">official documentation in the NBA</a>.</p>
<p>All specimen occurrence services are accessible using the methods within the <code>SpecimenClient</code> class. For a list of available endpoints, please refer to the class documentation (<code><a href="../reference/Specimen.html">?Specimen</a></code>). Below, we will give details about all services, grouped by category.</p>
<div id="query-services" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services" class="anchor"></a>Query services</h3>
<div id="query" class="section level4">
<h4 class="hasAnchor">
<a href="#query" class="anchor"></a>query</h4>
<p>Querying for specimens is accomplished with the <code>query</code> method in the class <code>SpecimenClient</code>. For simple queries, query parameters of type <code>list</code> can be passed via the parameter <code>queryParams</code>, for example we can query specimens of the Family <em>Ebenaceae</em> that were collected in Europe:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="co"># instantiate specimen client</span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2">sc &lt;-<span class="st"> </span>SpecimenClient<span class="op">$</span><span class="kw">new</span>()</a>
<a class="sourceLine" id="cb13-3" data-line-number="3"></a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="co"># specify query params in named list</span></a>
<a class="sourceLine" id="cb13-5" data-line-number="5">qp &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">identifications.defaultClassification.family=</span><span class="st">"Ebenaceae"</span>, <span class="dt">gatheringEvent.continent=</span><span class="st">"Europe"</span>)</a>
<a class="sourceLine" id="cb13-6" data-line-number="6"></a>
<a class="sourceLine" id="cb13-7" data-line-number="7"><span class="co"># query</span></a>
<a class="sourceLine" id="cb13-8" data-line-number="8">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">query</span>(<span class="dt">queryParams=</span>qp)</a></code></pre></div>
<p>If we now want to know all the countries that the specimens were collected in, we can access the <code>Specimen</code> objects in <code>res$content$resultSet</code> as follows:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">sapply</span>(res<span class="op">$</span>content<span class="op">$</span>resultSet, <span class="cf">function</span>(x)x<span class="op">$</span>item<span class="op">$</span>gatheringEvent<span class="op">$</span>country)</a></code></pre></div>
<pre><code>##  [1] "Italy"       "Austria"     "Switzerland" "Netherlands" "Netherlands"
##  [6] "Switzerland" "Italy"       "Netherlands" "France"      "Austria"</code></pre>
<p>Note that passing query parameters as a named list only allows for limited queries; the logical conjunction between parameters is for example always <code>AND</code>. More complex queries can be accomplished using the <code>QuerySpec</code> object:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="co"># get all specimens with genus name starting with 'Hydro'</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"identifications.defaultClassification.genus"</span>, <span class="dt">operator=</span><span class="st">"STARTS_WITH"</span>, <span class="dt">value=</span><span class="st">"Hydro"</span>)</a>
<a class="sourceLine" id="cb16-3" data-line-number="3">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(qc))</a>
<a class="sourceLine" id="cb16-4" data-line-number="4">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">query</span>(qs)</a></code></pre></div>
</div>
<div id="download_query" class="section level4">
<h4 class="hasAnchor">
<a href="#download_query" class="anchor"></a>download_query</h4>
<p>The <code>query</code> function is limited to retrieve 50000 specimen at once (this is determined in the parameter <code>index.max_result_window</code>, the value is retrievable using the <code>getSettings</code> method in the metadata section). In order to provide access for a larger amount of data, the <code>query_download</code> takes the same arguments as <code>query</code>, but download the data as a gzip stream under the hood. Unlike <code>query</code>, <code>query_download</code> returns a list of specimen objects instead of a <code>ResultSet</code>. Example:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">## get the first 100000 specimen objects (not possible with query method)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">download_query</span>(QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">size=</span><span class="dv">100000</span>))</a></code></pre></div>
</div>
</div>
<div id="data-access-services" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services" class="anchor"></a>Data access services</h3>
<p>Several access methods offer the convenient retrieval of specimens matching a certain identifier or being part of a certain collection. Below we give examples of how to use the currently implemented data access services for specimen records:</p>
<div id="access-to-collections" class="section level4">
<h4 class="hasAnchor">
<a href="#access-to-collections" class="anchor"></a>Access to collections</h4>
<p>Some of the specimens available via the NBA are categorised thematically into special collections, such as the Siebold-, Dubois- or Jongmans collection. The function <code>get_named_collection</code> lists all available special collections and the identifiers of the specimens within a collection can be queries with <code>get_ids_in_collection</code></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_named_collections</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "jongmans"      "dubois"        "timor"         "mammoth"      
##  [5] "siebold"       "martin"        "arts"          "meteorites"   
##  [9] "en_tibi"       "extinct-birds"</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_ids_in_collection</span>(<span class="st">"siebold"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "L.1670559@BRAHMS"  "L  0834597@BRAHMS" "L  0834879@BRAHMS"
##  [4] "L  0834880@BRAHMS" "L  0246491@BRAHMS" "L  0246496@BRAHMS"
##  [7] "L  0325027@BRAHMS" "L  0325034@BRAHMS" "L  0325063@BRAHMS"
## [10] "L  0325099@BRAHMS"</code></pre>
</div>
<div id="count" class="section level4">
<h4 class="hasAnchor">
<a href="#count" class="anchor"></a>count</h4>
<p>For any given query (with <code>QuerySpec</code> or not), returns the count of matches instead of specimen objects:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co"># Example with QuerySpec:</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="co"># how many specimens are there in the 'Botany' collection?</span></a>
<a class="sourceLine" id="cb22-3" data-line-number="3">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">'collectionType'</span>, <span class="dt">operator=</span><span class="st">'EQUALS'</span>, <span class="dt">value=</span><span class="st">'Botany'</span>)</a>
<a class="sourceLine" id="cb22-4" data-line-number="4">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(qc))</a>
<a class="sourceLine" id="cb22-5" data-line-number="5"></a>
<a class="sourceLine" id="cb22-6" data-line-number="6"><span class="co"># get the number of specimens</span></a>
<a class="sourceLine" id="cb22-7" data-line-number="7">sc<span class="op">$</span><span class="kw">count</span>(qs)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] 4892987</code></pre>
<p>Note that the count of matches for a given query is also returned by the <code>query</code> function. However, <code>count</code> is more lightweight as it returns an <code>integer</code> instead of a <code>ResultSet</code> containing <code>Specimen</code> objects.</p>
</div>
<div id="exists" class="section level4">
<h4 class="hasAnchor">
<a href="#exists" class="anchor"></a>exists</h4>
<p>Check if a record exists, based on its <em>unitID</em>:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="co"># use SpecimenClient instantiated above</span></a>
<a class="sourceLine" id="cb24-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">exists</span>(<span class="st">'ZMA.INS.1255440'</span>)                                        </a>
<a class="sourceLine" id="cb24-3" data-line-number="3"></a>
<a class="sourceLine" id="cb24-4" data-line-number="4"><span class="co"># content is boolean</span></a>
<a class="sourceLine" id="cb24-5" data-line-number="5">res<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="find" class="section level4">
<h4 class="hasAnchor">
<a href="#find" class="anchor"></a>find</h4>
<p>Return a single specimen given its identifier (Note: the identifier of a specimen is different from the unitID, see also <a href="http://docs.biodiversitydata.nl/en/latest/doc-spec-services/specimen/#specids">here</a>):</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">id &lt;-<span class="st"> "RMNH.MAM.17209.B@CRS"</span></a>
<a class="sourceLine" id="cb26-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">find</span>(id)</a>
<a class="sourceLine" id="cb26-3" data-line-number="3"></a>
<a class="sourceLine" id="cb26-4" data-line-number="4"><span class="co"># content is single specimen object</span></a>
<a class="sourceLine" id="cb26-5" data-line-number="5">res<span class="op">$</span>content</a></code></pre></div>
<pre><code>## &lt;Specimen&gt;
##   Public:
##     acquiredFrom: Agent, R6
##     assemblageID: NULL
##     associatedMultiMediaUris: list
##     clone: function (deep = FALSE) 
##     collectionType: Mammalia
##     collectorsFieldNumber: NULL
##     fromCaptivity: FALSE
##     fromJSONString: function (SpecimenJson, typeMapping = NULL) 
##     fromList: function (SpecimenList, typeMapping = NULL) 
##     gatheringEvent: GatheringEvent, R6
##     id: RMNH.MAM.17209.B@CRS
##     identifications: list
##     initialize: function (sourceSystem, sourceSystemId, recordURI, id, unitID, 
##     kindOfUnit: skin
##     license: CC0
##     licenseType: Copyright
##     multiMediaPublic: TRUE
##     notes: NULL
##     numberOfSpecimen: 1
##     objectPublic: TRUE
##     owner: Naturalis Biodiversity Center
##     phaseOrStage: juvenile
##     preparationType: NULL
##     previousSourceID: NULL
##     previousUnitsText: NULL
##     recordBasis: PreservedSpecimen
##     recordURI: NULL
##     sex: female
##     sourceID: CRS
##     sourceInstitutionID: Naturalis Biodiversity Center
##     sourceSystem: SourceSystem, R6
##     sourceSystemId: RMNH.MAM.17209.B
##     theme: NULL
##     title: RMNH.MAM.17209.b_RMNH_MAM_17209.b
##     toJSONString: function (pretty = TRUE) 
##     toList: function () 
##     unitGUID: http://data.biodiversitydata.nl/naturalis/specimen/RMNH. ...
##     unitID: RMNH.MAM.17209.B</code></pre>
</div>
<div id="find_by_ids" class="section level4">
<h4 class="hasAnchor">
<a href="#find_by_ids" class="anchor"></a>find_by_ids</h4>
<p>Same as <code>find</code>, but takes multiple IDs:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">ids &lt;-<span class="st"> "RMNH.INS.657083@CRS,L.1589244@BRAHMS"</span></a>
<a class="sourceLine" id="cb28-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">find_by_ids</span>(ids)</a></code></pre></div>
</div>
<div id="find_by_unit_id" class="section level4">
<h4 class="hasAnchor">
<a href="#find_by_unit_id" class="anchor"></a>find_by_unit_id</h4>
<p>Find a specimen by its <code>unitID</code>:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1">unitID &lt;-<span class="st"> "RMNH.MAM.1513"</span></a>
<a class="sourceLine" id="cb29-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">find_by_unit_id</span>(unitID)</a></code></pre></div>
</div>
</div>
<div id="aggregation-services" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services" class="anchor"></a>Aggregation services</h3>
<p>Aggregation services group available data according to different criteria.</p>
<div id="get_distinct_values" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values" class="anchor"></a>get_distinct_values</h4>
<p>This method takes a specific field as an argument and returns all possible values and the frequency for that field in thy data. Below we get all possible values for the country in which a specimen was collected.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"gatheringEvent.country"</span>)</a></code></pre></div>
<p><strong>Note</strong>: By default, <code>get_distinct_values</code> lists only the first 10 hits. The above query thus does not reflect the distinct values in the hole dataset. This number can be increased with e.g. setting the <code>size</code> parameter in a <code>QuerySpec</code> object passed to the method.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"gatheringEvent.country"</span>, <span class="dt">querySpec=</span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">size=</span><span class="dv">10000</span>))</a></code></pre></div>
</div>
<div id="count_distinct_values" class="section level4">
<h4 class="hasAnchor">
<a href="#count_distinct_values" class="anchor"></a>count_distinct_values</h4>
<p>Instead of returning all different values for a given field, this method does a mere count:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">sc<span class="op">$</span><span class="kw">count_distinct_values</span>(<span class="st">"gatheringEvent.country"</span>)</a></code></pre></div>
</div>
<div id="get_distinct_values_per_group" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values_per_group" class="anchor"></a>get_distinct_values_per_group</h4>
<p>Suppose you want to add another filter to the above query of retrieving all distinct values for a given field, such as splitting the results by their respective source system:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_distinct_values_per_group</span>(<span class="st">"sourceSystem.name"</span>, <span class="st">"gatheringEvent.country"</span>)</a></code></pre></div>
</div>
<div id="count_distinct_values_per_group" class="section level4">
<h4 class="hasAnchor">
<a href="#count_distinct_values_per_group" class="anchor"></a>count_distinct_values_per_group</h4>
<p>If no return of the actual values is needed, a simple count per group is done as follows:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1">sc<span class="op">$</span><span class="kw">count_distinct_values_per_group</span>(<span class="st">"sourceSystem.name"</span>, <span class="st">"gatheringEvent.country"</span>)<span class="op">$</span>content</a></code></pre></div>
</div>
</div>
<div id="metadata-services" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services" class="anchor"></a>Metadata services</h3>
<p>Specimen Metadata services include the same standard metadata services as for the other data types:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1"><span class="co"># get all paths for the Specimen datatype </span></a>
<a class="sourceLine" id="cb35-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##   [1] "sourceSystem.code"                                                       
##   [2] "sourceSystem.name"                                                       
##   [3] "sourceSystemId"                                                          
##   [4] "recordURI"                                                               
##   [5] "unitID"                                                                  
##   [6] "unitGUID"                                                                
##   [7] "collectorsFieldNumber"                                                   
##   [8] "assemblageID"                                                            
##   [9] "sourceInstitutionID"                                                     
##  [10] "sourceID"                                                                
##  [11] "previousSourceID"                                                        
##  [12] "owner"                                                                   
##  [13] "licenseType"                                                             
##  [14] "license"                                                                 
##  [15] "recordBasis"                                                             
##  [16] "kindOfUnit"                                                              
##  [17] "collectionType"                                                          
##  [18] "sex"                                                                     
##  [19] "phaseOrStage"                                                            
##  [20] "title"                                                                   
##  [21] "notes"                                                                   
##  [22] "preparationType"                                                         
##  [23] "previousUnitsText"                                                       
##  [24] "numberOfSpecimen"                                                        
##  [25] "fromCaptivity"                                                           
##  [26] "objectPublic"                                                            
##  [27] "multiMediaPublic"                                                        
##  [28] "acquiredFrom.agentText"                                                  
##  [29] "gatheringEvent.projectTitle"                                             
##  [30] "gatheringEvent.worldRegion"                                              
##  [31] "gatheringEvent.continent"                                                
##  [32] "gatheringEvent.country"                                                  
##  [33] "gatheringEvent.iso3166Code"                                              
##  [34] "gatheringEvent.provinceState"                                            
##  [35] "gatheringEvent.island"                                                   
##  [36] "gatheringEvent.locality"                                                 
##  [37] "gatheringEvent.city"                                                     
##  [38] "gatheringEvent.sublocality"                                              
##  [39] "gatheringEvent.localityText"                                             
##  [40] "gatheringEvent.dateTimeBegin"                                            
##  [41] "gatheringEvent.dateTimeEnd"                                              
##  [42] "gatheringEvent.method"                                                   
##  [43] "gatheringEvent.altitude"                                                 
##  [44] "gatheringEvent.altitudeUnifOfMeasurement"                                
##  [45] "gatheringEvent.biotopeText"                                              
##  [46] "gatheringEvent.depth"                                                    
##  [47] "gatheringEvent.depthUnitOfMeasurement"                                   
##  [48] "gatheringEvent.gatheringPersons.agentText"                               
##  [49] "gatheringEvent.gatheringPersons.fullName"                                
##  [50] "gatheringEvent.gatheringPersons.organization.agentText"                  
##  [51] "gatheringEvent.gatheringPersons.organization.name"                       
##  [52] "gatheringEvent.gatheringOrganizations.agentText"                         
##  [53] "gatheringEvent.gatheringOrganizations.name"                              
##  [54] "gatheringEvent.siteCoordinates.longitudeDecimal"                         
##  [55] "gatheringEvent.siteCoordinates.latitudeDecimal"                          
##  [56] "gatheringEvent.siteCoordinates.gridCellSystem"                           
##  [57] "gatheringEvent.siteCoordinates.gridLatitudeDecimal"                      
##  [58] "gatheringEvent.siteCoordinates.gridLongitudeDecimal"                     
##  [59] "gatheringEvent.siteCoordinates.gridCellCode"                             
##  [60] "gatheringEvent.siteCoordinates.gridQualifier"                            
##  [61] "gatheringEvent.siteCoordinates.geoShape"                                 
##  [62] "gatheringEvent.namedAreas.areaName"                                      
##  [63] "gatheringEvent.namedAreas.areaClass"                                     
##  [64] "gatheringEvent.associatedTaxa.name"                                      
##  [65] "gatheringEvent.associatedTaxa.relationType"                              
##  [66] "gatheringEvent.chronoStratigraphy.youngRegionalSubstage"                 
##  [67] "gatheringEvent.chronoStratigraphy.youngRegionalStage"                    
##  [68] "gatheringEvent.chronoStratigraphy.youngRegionalSeries"                   
##  [69] "gatheringEvent.chronoStratigraphy.youngDatingQualifier"                  
##  [70] "gatheringEvent.chronoStratigraphy.youngInternSystem"                     
##  [71] "gatheringEvent.chronoStratigraphy.youngInternSubstage"                   
##  [72] "gatheringEvent.chronoStratigraphy.youngInternStage"                      
##  [73] "gatheringEvent.chronoStratigraphy.youngInternSeries"                     
##  [74] "gatheringEvent.chronoStratigraphy.youngInternErathem"                    
##  [75] "gatheringEvent.chronoStratigraphy.youngInternEonothem"                   
##  [76] "gatheringEvent.chronoStratigraphy.youngChronoName"                       
##  [77] "gatheringEvent.chronoStratigraphy.youngCertainty"                        
##  [78] "gatheringEvent.chronoStratigraphy.oldDatingQualifier"                    
##  [79] "gatheringEvent.chronoStratigraphy.chronoPreferredFlag"                   
##  [80] "gatheringEvent.chronoStratigraphy.oldRegionalSubstage"                   
##  [81] "gatheringEvent.chronoStratigraphy.oldRegionalStage"                      
##  [82] "gatheringEvent.chronoStratigraphy.oldRegionalSeries"                     
##  [83] "gatheringEvent.chronoStratigraphy.oldInternSystem"                       
##  [84] "gatheringEvent.chronoStratigraphy.oldInternSubstage"                     
##  [85] "gatheringEvent.chronoStratigraphy.oldInternStage"                        
##  [86] "gatheringEvent.chronoStratigraphy.oldInternSeries"                       
##  [87] "gatheringEvent.chronoStratigraphy.oldInternErathem"                      
##  [88] "gatheringEvent.chronoStratigraphy.oldInternEonothem"                     
##  [89] "gatheringEvent.chronoStratigraphy.oldChronoName"                         
##  [90] "gatheringEvent.chronoStratigraphy.chronoIdentifier"                      
##  [91] "gatheringEvent.chronoStratigraphy.oldCertainty"                          
##  [92] "gatheringEvent.bioStratigraphy.youngBioDatingQualifier"                  
##  [93] "gatheringEvent.bioStratigraphy.youngBioName"                             
##  [94] "gatheringEvent.bioStratigraphy.youngFossilZone"                          
##  [95] "gatheringEvent.bioStratigraphy.youngFossilSubZone"                       
##  [96] "gatheringEvent.bioStratigraphy.youngBioCertainty"                        
##  [97] "gatheringEvent.bioStratigraphy.youngStratType"                           
##  [98] "gatheringEvent.bioStratigraphy.bioDatingQualifier"                       
##  [99] "gatheringEvent.bioStratigraphy.bioPreferredFlag"                         
## [100] "gatheringEvent.bioStratigraphy.rangePosition"                            
## [101] "gatheringEvent.bioStratigraphy.oldBioName"                               
## [102] "gatheringEvent.bioStratigraphy.bioIdentifier"                            
## [103] "gatheringEvent.bioStratigraphy.oldFossilzone"                            
## [104] "gatheringEvent.bioStratigraphy.oldFossilSubzone"                         
## [105] "gatheringEvent.bioStratigraphy.oldBioCertainty"                          
## [106] "gatheringEvent.bioStratigraphy.oldBioStratType"                          
## [107] "gatheringEvent.lithoStratigraphy.qualifier"                              
## [108] "gatheringEvent.lithoStratigraphy.preferredFlag"                          
## [109] "gatheringEvent.lithoStratigraphy.member2"                                
## [110] "gatheringEvent.lithoStratigraphy.member"                                 
## [111] "gatheringEvent.lithoStratigraphy.informalName2"                          
## [112] "gatheringEvent.lithoStratigraphy.informalName"                           
## [113] "gatheringEvent.lithoStratigraphy.importedName2"                          
## [114] "gatheringEvent.lithoStratigraphy.importedName1"                          
## [115] "gatheringEvent.lithoStratigraphy.lithoIdentifier"                        
## [116] "gatheringEvent.lithoStratigraphy.formation2"                             
## [117] "gatheringEvent.lithoStratigraphy.formationGroup2"                        
## [118] "gatheringEvent.lithoStratigraphy.formationGroup"                         
## [119] "gatheringEvent.lithoStratigraphy.formation"                              
## [120] "gatheringEvent.lithoStratigraphy.certainty2"                             
## [121] "gatheringEvent.lithoStratigraphy.certainty"                              
## [122] "gatheringEvent.lithoStratigraphy.bed2"                                   
## [123] "gatheringEvent.lithoStratigraphy.bed"                                    
## [124] "identifications.taxonRank"                                               
## [125] "identifications.scientificName.fullScientificName"                       
## [126] "identifications.scientificName.taxonomicStatus"                          
## [127] "identifications.scientificName.genusOrMonomial"                          
## [128] "identifications.scientificName.subgenus"                                 
## [129] "identifications.scientificName.specificEpithet"                          
## [130] "identifications.scientificName.infraspecificEpithet"                     
## [131] "identifications.scientificName.infraspecificMarker"                      
## [132] "identifications.scientificName.nameAddendum"                             
## [133] "identifications.scientificName.authorshipVerbatim"                       
## [134] "identifications.scientificName.author"                                   
## [135] "identifications.scientificName.year"                                     
## [136] "identifications.scientificName.scientificNameGroup"                      
## [137] "identifications.scientificName.references.titleCitation"                 
## [138] "identifications.scientificName.references.citationDetail"                
## [139] "identifications.scientificName.references.uri"                           
## [140] "identifications.scientificName.references.author.agentText"              
## [141] "identifications.scientificName.references.author.fullName"               
## [142] "identifications.scientificName.references.author.organization.agentText" 
## [143] "identifications.scientificName.references.author.organization.name"      
## [144] "identifications.scientificName.references.publicationDate"               
## [145] "identifications.scientificName.experts.agentText"                        
## [146] "identifications.scientificName.experts.fullName"                         
## [147] "identifications.scientificName.experts.organization.agentText"           
## [148] "identifications.scientificName.experts.organization.name"                
## [149] "identifications.typeStatus"                                              
## [150] "identifications.dateIdentified"                                          
## [151] "identifications.defaultClassification.kingdom"                           
## [152] "identifications.defaultClassification.phylum"                            
## [153] "identifications.defaultClassification.className"                         
## [154] "identifications.defaultClassification.order"                             
## [155] "identifications.defaultClassification.superFamily"                       
## [156] "identifications.defaultClassification.family"                            
## [157] "identifications.defaultClassification.genus"                             
## [158] "identifications.defaultClassification.subgenus"                          
## [159] "identifications.defaultClassification.specificEpithet"                   
## [160] "identifications.defaultClassification.infraspecificEpithet"              
## [161] "identifications.defaultClassification.infraspecificRank"                 
## [162] "identifications.systemClassification.rank"                               
## [163] "identifications.systemClassification.name"                               
## [164] "identifications.vernacularNames.name"                                    
## [165] "identifications.vernacularNames.language"                                
## [166] "identifications.vernacularNames.preferred"                               
## [167] "identifications.vernacularNames.references.titleCitation"                
## [168] "identifications.vernacularNames.references.citationDetail"               
## [169] "identifications.vernacularNames.references.uri"                          
## [170] "identifications.vernacularNames.references.author.agentText"             
## [171] "identifications.vernacularNames.references.author.fullName"              
## [172] "identifications.vernacularNames.references.author.organization.agentText"
## [173] "identifications.vernacularNames.references.author.organization.name"     
## [174] "identifications.vernacularNames.references.publicationDate"              
## [175] "identifications.vernacularNames.experts.agentText"                       
## [176] "identifications.vernacularNames.experts.fullName"                        
## [177] "identifications.vernacularNames.experts.organization.agentText"          
## [178] "identifications.vernacularNames.experts.organization.name"               
## [179] "identifications.identificationQualifiers"                                
## [180] "identifications.identifiers.agentText"                                   
## [181] "identifications.taxonomicEnrichments.vernacularNames.name"               
## [182] "identifications.taxonomicEnrichments.vernacularNames.language"           
## [183] "identifications.taxonomicEnrichments.synonyms.fullScientificName"        
## [184] "identifications.taxonomicEnrichments.synonyms.taxonomicStatus"           
## [185] "identifications.taxonomicEnrichments.synonyms.genusOrMonomial"           
## [186] "identifications.taxonomicEnrichments.synonyms.subgenus"                  
## [187] "identifications.taxonomicEnrichments.synonyms.specificEpithet"           
## [188] "identifications.taxonomicEnrichments.synonyms.infraspecificEpithet"      
## [189] "identifications.taxonomicEnrichments.synonyms.authorshipVerbatim"        
## [190] "identifications.taxonomicEnrichments.sourceSystem.code"                  
## [191] "identifications.taxonomicEnrichments.taxonId"                            
## [192] "identifications.preferred"                                               
## [193] "identifications.verificationStatus"                                      
## [194] "identifications.rockType"                                                
## [195] "identifications.associatedFossilAssemblage"                              
## [196] "identifications.rockMineralUsage"                                        
## [197] "identifications.associatedMineralName"                                   
## [198] "identifications.remarks"                                                 
## [199] "associatedMultiMediaUris.accessUri"                                      
## [200] "associatedMultiMediaUris.format"                                         
## [201] "associatedMultiMediaUris.variant"                                        
## [202] "theme"</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1"><span class="co"># get info e.g. for field collectionType</span></a>
<a class="sourceLine" id="cb37-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content<span class="op">$</span>collectionType</a></code></pre></div>
<pre><code>## $indexed
## [1] TRUE
## 
## $type
## [1] "keyword"
## 
## $allowedOperators
##  [1] "EQUALS"             "NOT_EQUALS"         "EQUALS_IC"         
##  [4] "NOT_EQUALS_IC"      "IN"                 "NOT_IN"            
##  [7] "MATCHES"            "NOT_MATCHES"        "STARTS_WITH"       
## [10] "NOT_STARTS_WITH"    "STARTS_WITH_IC"     "NOT_STARTS_WITH_IC"</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1"><span class="co"># get all settings</span></a>
<a class="sourceLine" id="cb39-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_settings</span>()</a></code></pre></div>
<pre><code>## &lt;Response&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     content: list
##     initialize: function (content, response) 
##     response: response</code></pre>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1"><span class="co"># get specific setting</span></a>
<a class="sourceLine" id="cb41-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1"><span class="co"># check if operator is allowed</span></a>
<a class="sourceLine" id="cb43-2" data-line-number="2">sc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"gatheringEvent.continent"</span>, <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div id="available-paths-and-fields" class="section level4">
<h4 class="hasAnchor">
<a href="#available-paths-and-fields" class="anchor"></a>Available paths and fields</h4>
<p>All fields can be retrieved with <code>get_paths</code> and specific information on the fields, such as allowed operators etc. with <code>get_field_info</code></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a>
<a class="sourceLine" id="cb45-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content</a></code></pre></div>
</div>
<div id="settings" class="section level4">
<h4 class="hasAnchor">
<a href="#settings" class="anchor"></a>Settings</h4>
<p>Multimedia-specific settings can be retrieved with <code>get_settings</code> and a specific setting with <code>get_setting</code>:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_settings</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $index.max_result_window
## [1] 50000
## 
## $specimen.group_by_scientific_name.max_num_buckets
## [1] 10000</code></pre>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
</div>
<div id="operators" class="section level4">
<h4 class="hasAnchor">
<a href="#operators" class="anchor"></a>Operators</h4>
<p>To test if a certain operator can be used for a multimedia query:</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" data-line-number="1">sc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"identifications.defaultClassification.genus"</span>, <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="dwca-download-services" class="section level3">
<h3 class="hasAnchor">
<a href="#dwca-download-services" class="anchor"></a>DwCA download services</h3>
<p>In addition to query services that return JSON formatted data, the NBA also offers the export of <a href="https://en.wikipedia.org/wiki/Darwin_Core_Archive">Darwin Core Archive (DwCA)</a> files. These files are by default zip archives, please refer to our <a href="https://en.wikipedia.org/wiki/Darwin_Core_Archive">official API documentation</a> for more information.</p>
<div id="static-download" class="section level4">
<h4 class="hasAnchor">
<a href="#static-download" class="anchor"></a>Static download</h4>
<p>Static download services offer the download of predefined datasets. The sets that are available for download can be queried with <code>dwca_get_data_set_names</code>:</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1">sc<span class="op">$</span><span class="kw">dwca_get_data_set_names</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "amphibia-and-reptilia"      "aves"                      
##  [3] "botany"                     "brachiopoda"               
##  [5] "cainozoic-mollusca"         "chelicerata-and-myriapoda" 
##  [7] "cnidaria"                   "coleoptera"                
##  [9] "collembola"                 "crustacea"                 
## [11] "diptera"                    "echinodermata"             
## [13] "foraminifera"               "hemiptera"                 
## [15] "hymenoptera"                "lepidoptera"               
## [17] "mammalia"                   "micropaleontology"         
## [19] "mollusca"                   "odonata"                   
## [21] "orthopteroidea"             "paleobotany"               
## [23] "paleontology-invertebrates" "pisces"                    
## [25] "porifera"                   "tunicata"</code></pre>
<p>A dataset can then be downloaded using <code>dwca_get_data_set</code>. A filename can be given as argument, if none is given, the DwCA archive is written to <em>download-YYYY-MM-DDThh:mm.zip</em> in the current working directory.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1"><span class="co"># download dataset 'porifera' to temporary file</span></a>
<a class="sourceLine" id="cb54-2" data-line-number="2">filename &lt;-<span class="st"> </span><span class="kw">tempfile</span>(<span class="dt">fileext=</span><span class="st">".zip"</span>)</a>
<a class="sourceLine" id="cb54-3" data-line-number="3">sc<span class="op">$</span><span class="kw">dwca_get_data_set</span>(<span class="st">'porifera'</span>, <span class="dt">filename=</span>filename)                                                                               </a></code></pre></div>
<pre><code>## 
Downloading: 8 kB     
Downloading: 8 kB     
Downloading: 16 kB     
Downloading: 16 kB     
Downloading: 16 kB     
Downloading: 16 kB     
Downloading: 26 kB     
Downloading: 26 kB     
Downloading: 34 kB     
Downloading: 34 kB     
Downloading: 48 kB     
Downloading: 48 kB     
Downloading: 49 kB     
Downloading: 49 kB     
Downloading: 49 kB     
Downloading: 49 kB     
Downloading: 53 kB     
Downloading: 53 kB     
Downloading: 70 kB     
Downloading: 70 kB     
Downloading: 73 kB     
Downloading: 73 kB     
Downloading: 77 kB     
Downloading: 77 kB     
Downloading: 82 kB     
Downloading: 82 kB     
Downloading: 82 kB     
Downloading: 82 kB     
Downloading: 88 kB     
Downloading: 88 kB     
Downloading: 100 kB     
Downloading: 100 kB     
Downloading: 110 kB     
Downloading: 110 kB     
Downloading: 120 kB     
Downloading: 120 kB     
Downloading: 120 kB     
Downloading: 120 kB     
Downloading: 120 kB     
Downloading: 120 kB     
Downloading: 130 kB     
Downloading: 130 kB     
Downloading: 130 kB     
Downloading: 130 kB     
Downloading: 140 kB     
Downloading: 140 kB     
Downloading: 160 kB     
Downloading: 160 kB     
Downloading: 160 kB     
Downloading: 160 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 190 kB     
Downloading: 190 kB     
Downloading: 200 kB     
Downloading: 200 kB     
Downloading: 200 kB     
Downloading: 200 kB     
Downloading: 210 kB     
Downloading: 210 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 240 kB     
Downloading: 240 kB     
Downloading: 240 kB     
Downloading: 240 kB     
Downloading: 250 kB     
Downloading: 250 kB     
Downloading: 250 kB     
Downloading: 250 kB     
Downloading: 250 kB     
Downloading: 250 kB     
Downloading: 260 kB     
Downloading: 260 kB     
Downloading: 270 kB     
Downloading: 270 kB     
Downloading: 270 kB     
Downloading: 270 kB     
Downloading: 280 kB     
Downloading: 280 kB     
Downloading: 290 kB     
Downloading: 290 kB     
Downloading: 290 kB     
Downloading: 290 kB     
Downloading: 310 kB     
Downloading: 310 kB     
Downloading: 320 kB     
Downloading: 320 kB     
Downloading: 320 kB     
Downloading: 320 kB     
Downloading: 330 kB     
Downloading: 330 kB     
Downloading: 350 kB     
Downloading: 350 kB     
Downloading: 350 kB     
Downloading: 350 kB     
Downloading: 350 kB     
Downloading: 350 kB     
Downloading: 360 kB     
Downloading: 360 kB     
Downloading: 370 kB     
Downloading: 370 kB     
Downloading: 370 kB     
Downloading: 370 kB     
Downloading: 370 kB     
Downloading: 370 kB     
Downloading: 380 kB     
Downloading: 380 kB     
Downloading: 380 kB     
Downloading: 380 kB     
Downloading: 400 kB     
Downloading: 400 kB     
Downloading: 400 kB     
Downloading: 400 kB     
Downloading: 400 kB     
Downloading: 400 kB     
Downloading: 400 kB     
Downloading: 400 kB     
Downloading: 410 kB     
Downloading: 410 kB     
Downloading: 430 kB     
Downloading: 430 kB     
Downloading: 430 kB     
Downloading: 430 kB     
Downloading: 430 kB     
Downloading: 430 kB     
Downloading: 440 kB     
Downloading: 440 kB     
Downloading: 450 kB     
Downloading: 450 kB     
Downloading: 470 kB     
Downloading: 470 kB     
Downloading: 470 kB     
Downloading: 470 kB     
Downloading: 470 kB     
Downloading: 470 kB     
Downloading: 470 kB     
Downloading: 470 kB     
Downloading: 490 kB     
Downloading: 490 kB     
Downloading: 490 kB     
Downloading: 490 kB     
Downloading: 490 kB     
Downloading: 490 kB     
Downloading: 500 kB     
Downloading: 500 kB     
Downloading: 520 kB     
Downloading: 520 kB     
Downloading: 520 kB     
Downloading: 520 kB     
Downloading: 520 kB     
Downloading: 520 kB     
Downloading: 540 kB     
Downloading: 540 kB     
Downloading: 550 kB     
Downloading: 550 kB     
Downloading: 550 kB     
Downloading: 550 kB     
Downloading: 570 kB     
Downloading: 570 kB     
Downloading: 580 kB     
Downloading: 580 kB     
Downloading: 590 kB     
Downloading: 590 kB     
Downloading: 590 kB     
Downloading: 590 kB     
Downloading: 600 kB     
Downloading: 600 kB     
Downloading: 600 kB     
Downloading: 600 kB     
Downloading: 600 kB     
Downloading: 600 kB     
Downloading: 610 kB     
Downloading: 610 kB     
Downloading: 620 kB     
Downloading: 620 kB     
Downloading: 630 kB     
Downloading: 630 kB     
Downloading: 630 kB     
Downloading: 630 kB     
Downloading: 640 kB     
Downloading: 640 kB     
Downloading: 660 kB     
Downloading: 660 kB     
Downloading: 670 kB     
Downloading: 670 kB     
Downloading: 670 kB     
Downloading: 670 kB     
Downloading: 670 kB     
Downloading: 670 kB     
Downloading: 690 kB     
Downloading: 690 kB     
Downloading: 700 kB     
Downloading: 700 kB     
Downloading: 700 kB     
Downloading: 700 kB     
Downloading: 700 kB     
Downloading: 700 kB     
Downloading: 720 kB     
Downloading: 720 kB     
Downloading: 720 kB     
Downloading: 720 kB     
Downloading: 720 kB     
Downloading: 720 kB     
Downloading: 740 kB     
Downloading: 740 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 760 kB     
Downloading: 760 kB     
Downloading: 780 kB     
Downloading: 780 kB     
Downloading: 780 kB     
Downloading: 780 kB     
Downloading: 790 kB     
Downloading: 790 kB     
Downloading: 790 kB     
Downloading: 790 kB     
Downloading: 800 kB     
Downloading: 800 kB     
Downloading: 800 kB     
Downloading: 800 kB     
Downloading: 810 kB     
Downloading: 810 kB     
Downloading: 830 kB     
Downloading: 830 kB     
Downloading: 840 kB     
Downloading: 840 kB     
Downloading: 840 kB     
Downloading: 840 kB     
Downloading: 850 kB     
Downloading: 850 kB     
Downloading: 870 kB     
Downloading: 870 kB     
Downloading: 870 kB     
Downloading: 870 kB     
Downloading: 870 kB     
Downloading: 870 kB     
Downloading: 870 kB     
Downloading: 870 kB     
Downloading: 890 kB     
Downloading: 890 kB     
Downloading: 900 kB     
Downloading: 900 kB     
Downloading: 900 kB     
Downloading: 900 kB     
Downloading: 910 kB     
Downloading: 910 kB     
Downloading: 920 kB     
Downloading: 920 kB     
Downloading: 920 kB     
Downloading: 920 kB     
Downloading: 920 kB     
Downloading: 920 kB     
Downloading: 940 kB     
Downloading: 940 kB     
Downloading: 950 kB     
Downloading: 950 kB     
Downloading: 950 kB     
Downloading: 950 kB     
Downloading: 960 kB     
Downloading: 960 kB     
Downloading: 980 kB     
Downloading: 980 kB     
Downloading: 990 kB     
Downloading: 990 kB     
Downloading: 990 kB     
Downloading: 990 kB     
Downloading: 990 kB     
Downloading: 990 kB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB</code></pre>
<pre><code>## Query result written to /tmp/RtmpBgaqDP/file49dd1b98405.zip</code></pre>
<pre><code>## &lt;Response&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     content: NULL
##     initialize: function (content, response) 
##     response: response</code></pre>
</div>
<div id="dynamic-download" class="section level4">
<h4 class="hasAnchor">
<a href="#dynamic-download" class="anchor"></a>Dynamic download</h4>
<p>The dynamic download function <code>dwca_query</code> allows for download of arbitrary sets, defined by the user’s query. The arguments to this methods are similar to <code>query</code>, plus the filename:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" data-line-number="1"><span class="co"># download all specimen of genus 'Hydrochoerus'</span></a>
<a class="sourceLine" id="cb58-2" data-line-number="2">filename &lt;-<span class="st"> </span><span class="kw">tempfile</span>(<span class="dt">fileext=</span><span class="st">".zip"</span>)</a>
<a class="sourceLine" id="cb58-3" data-line-number="3">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"identifications.defaultClassification.genus"</span>, <span class="dt">operator=</span><span class="st">"EQUALS"</span>, <span class="dt">value=</span><span class="st">"Hydrochoerus"</span>)))</a>
<a class="sourceLine" id="cb58-4" data-line-number="4">sc<span class="op">$</span><span class="kw">dwca_query</span>(<span class="dt">querySpec=</span>qs, <span class="dt">filename=</span>filename)</a></code></pre></div>
<pre><code>## 
Downloading: 2.7 kB     
Downloading: 2.7 kB     
Downloading: 2.7 kB     
Downloading: 2.7 kB</code></pre>
<pre><code>## Query result written to /tmp/RtmpBgaqDP/file49dd5b8ad483.zip</code></pre>
<pre><code>## &lt;Response&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     content: NULL
##     initialize: function (content, response) 
##     response: response</code></pre>
</div>
</div>
</div>
<div id="taxonomic-data-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#taxonomic-data-services" class="anchor"></a>Taxonomic data services</h2>
<div id="query-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services-1" class="anchor"></a>Query Services</h3>
<div id="query-1" class="section level4">
<h4 class="hasAnchor">
<a href="#query-1" class="anchor"></a>query</h4>
<p>Query for taxon document with given search criteria. Example:</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb62-1" data-line-number="1"><span class="co"># query for taxa of genus 'Sedum' that are in the Netherlands Soortenregister (NSR)</span></a>
<a class="sourceLine" id="cb62-2" data-line-number="2">tc &lt;-<span class="st"> </span>TaxonClient<span class="op">$</span><span class="kw">new</span>()</a>
<a class="sourceLine" id="cb62-3" data-line-number="3">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"acceptedName.genusOrMonomial"</span>, <span class="dt">operator=</span><span class="st">"EQUALS"</span>, <span class="dt">value=</span><span class="st">"Sedum"</span>)</a>
<a class="sourceLine" id="cb62-4" data-line-number="4">qc2 &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"sourceSystem.code"</span>, <span class="dt">operator=</span><span class="st">"EQUALS"</span>, <span class="dt">value=</span><span class="st">"NSR"</span>)</a>
<a class="sourceLine" id="cb62-5" data-line-number="5">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(qc, qc2))</a>
<a class="sourceLine" id="cb62-6" data-line-number="6">tc<span class="op">$</span><span class="kw">query</span>(qs)</a></code></pre></div>
</div>
<div id="download_query-1" class="section level4">
<h4 class="hasAnchor">
<a href="#download_query-1" class="anchor"></a>download_query</h4>
<p>The <code>query</code> function is limited to retrieve 50000 taxa at once (this is determined in the parameter <code>index.max_result_window</code>, the value is retrievable using the <code>getSettings</code> method in the metadata section). In order to provide access for a larger amount of data, the <code>query_download</code> takes the same arguments as <code>query</code>, but download the data as a gzip stream under the hood. Unlike <code>query</code>, <code>query_download</code> returns a list of taxon objects instead of a <code>ResultSet</code>.</p>
</div>
</div>
<div id="data-access-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services-1" class="anchor"></a>Data access services</h3>
<div id="count-1" class="section level4">
<h4 class="hasAnchor">
<a href="#count-1" class="anchor"></a>count</h4>
<p>For a given query, do not return <code>Taxon</code> objects but the mere count. Example</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" data-line-number="1"><span class="co"># get counts for taxa of genus 'Sedum' that are in the Netherlands Soortenregister (NSR)</span></a>
<a class="sourceLine" id="cb63-2" data-line-number="2">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"acceptedName.genusOrMonomial"</span>, <span class="dt">operator=</span><span class="st">"EQUALS"</span>, <span class="dt">value=</span><span class="st">"Sedum"</span>)</a>
<a class="sourceLine" id="cb63-3" data-line-number="3">qc2 &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"sourceSystem.code"</span>, <span class="dt">operator=</span><span class="st">"EQUALS"</span>, <span class="dt">value=</span><span class="st">"NSR"</span>)</a>
<a class="sourceLine" id="cb63-4" data-line-number="4">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(qc, qc2))</a>
<a class="sourceLine" id="cb63-5" data-line-number="5">tc<span class="op">$</span><span class="kw">count</span>(qs)</a></code></pre></div>
<pre><code>## &lt;Response&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     content: 23
##     initialize: function (content, response) 
##     response: response</code></pre>
</div>
<div id="find-1" class="section level4">
<h4 class="hasAnchor">
<a href="#find-1" class="anchor"></a>find</h4>
<p>Returns a taxon object given its identifier:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" data-line-number="1">tc<span class="op">$</span><span class="kw">find</span>(<span class="st">"27706109@COL"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## &lt;Taxon&gt;
##   Public:
##     acceptedName: ScientificName, R6
##     clone: function (deep = FALSE) 
##     defaultClassification: DefaultClassification, R6
##     descriptions: list
##     experts: list
##     fromJSONString: function (TaxonJson, typeMapping = NULL) 
##     fromList: function (TaxonList, typeMapping = NULL) 
##     id: 27706109@COL
##     initialize: function (sourceSystem, sourceSystemId, recordURI, id, sourceSystemParentId, 
##     occurrenceStatusVerbatim: NULL
##     recordURI: http://www.catalogueoflife.org/annual-checklist/2017/det ...
##     references: list
##     sourceSystem: SourceSystem, R6
##     sourceSystemId: 27706109
##     sourceSystemParentId: NULL
##     synonyms: list
##     systemClassification: list
##     taxonRank: infraspecies
##     taxonRemarks: NULL
##     toJSONString: function (pretty = TRUE) 
##     toList: function () 
##     validName: ScientificName, R6
##     vernacularNames: list</code></pre>
</div>
<div id="find_by_ids-1" class="section level4">
<h4 class="hasAnchor">
<a href="#find_by_ids-1" class="anchor"></a>find_by_ids</h4>
<p>Given a string with comma-separated identifiers, returns a list of taxon objects:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" data-line-number="1">    ids &lt;-<span class="st"> "27706109@COL,27704140@COL,27706110@COL,27706111@COL,27706108@COL"</span></a>
<a class="sourceLine" id="cb67-2" data-line-number="2">    res &lt;-<span class="st"> </span>tc<span class="op">$</span><span class="kw">find_by_ids</span>(ids)</a></code></pre></div>
</div>
</div>
<div id="aggregation-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services-1" class="anchor"></a>Aggregation services</h3>
<div id="get_distinct_values-1" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values-1" class="anchor"></a>get_distinct_values</h4>
<p>This method takes a specific field as an argument and returns all possible values and the frequency for that field in thy data. Example: get all data source systems for taxon objects:</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb68-1" data-line-number="1">tc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"sourceSystem.name"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $`Species 2000 - Catalogue Of Life`
## [1] 1865315
## 
## $`Naturalis - Dutch Species Register`
## [1] 47611</code></pre>
</div>
</div>
<div id="metadata-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services-1" class="anchor"></a>Metadata services</h3>
<p>Taxon Metadata services include the same standard metadata services as for the other data types:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb70-1" data-line-number="1"><span class="co"># get all paths for the Taxon datatype </span></a>
<a class="sourceLine" id="cb70-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "sourceSystem.code"                                       
##  [2] "sourceSystem.name"                                       
##  [3] "sourceSystemId"                                          
##  [4] "recordURI"                                               
##  [5] "sourceSystemParentId"                                    
##  [6] "taxonRank"                                               
##  [7] "taxonRemarks"                                            
##  [8] "occurrenceStatusVerbatim"                                
##  [9] "acceptedName.fullScientificName"                         
## [10] "acceptedName.taxonomicStatus"                            
## [11] "acceptedName.genusOrMonomial"                            
## [12] "acceptedName.subgenus"                                   
## [13] "acceptedName.specificEpithet"                            
## [14] "acceptedName.infraspecificEpithet"                       
## [15] "acceptedName.infraspecificMarker"                        
## [16] "acceptedName.nameAddendum"                               
## [17] "acceptedName.authorshipVerbatim"                         
## [18] "acceptedName.author"                                     
## [19] "acceptedName.year"                                       
## [20] "acceptedName.scientificNameGroup"                        
## [21] "acceptedName.references.titleCitation"                   
## [22] "acceptedName.references.citationDetail"                  
## [23] "acceptedName.references.uri"                             
## [24] "acceptedName.references.author.agentText"                
## [25] "acceptedName.references.author.fullName"                 
## [26] "acceptedName.references.author.organization.agentText"   
## [27] "acceptedName.references.author.organization.name"        
## [28] "acceptedName.references.publicationDate"                 
## [29] "acceptedName.experts.agentText"                          
## [30] "acceptedName.experts.fullName"                           
## [31] "acceptedName.experts.organization.agentText"             
## [32] "acceptedName.experts.organization.name"                  
## [33] "defaultClassification.kingdom"                           
## [34] "defaultClassification.phylum"                            
## [35] "defaultClassification.className"                         
## [36] "defaultClassification.order"                             
## [37] "defaultClassification.superFamily"                       
## [38] "defaultClassification.family"                            
## [39] "defaultClassification.genus"                             
## [40] "defaultClassification.subgenus"                          
## [41] "defaultClassification.specificEpithet"                   
## [42] "defaultClassification.infraspecificEpithet"              
## [43] "defaultClassification.infraspecificRank"                 
## [44] "systemClassification.rank"                               
## [45] "systemClassification.name"                               
## [46] "synonyms.fullScientificName"                             
## [47] "synonyms.taxonomicStatus"                                
## [48] "synonyms.genusOrMonomial"                                
## [49] "synonyms.subgenus"                                       
## [50] "synonyms.specificEpithet"                                
## [51] "synonyms.infraspecificEpithet"                           
## [52] "synonyms.infraspecificMarker"                            
## [53] "synonyms.nameAddendum"                                   
## [54] "synonyms.authorshipVerbatim"                             
## [55] "synonyms.author"                                         
## [56] "synonyms.year"                                           
## [57] "synonyms.scientificNameGroup"                            
## [58] "synonyms.references.titleCitation"                       
## [59] "synonyms.references.citationDetail"                      
## [60] "synonyms.references.uri"                                 
## [61] "synonyms.references.author.agentText"                    
## [62] "synonyms.references.author.fullName"                     
## [63] "synonyms.references.author.organization.agentText"       
## [64] "synonyms.references.author.organization.name"            
## [65] "synonyms.references.publicationDate"                     
## [66] "synonyms.experts.agentText"                              
## [67] "synonyms.experts.fullName"                               
## [68] "synonyms.experts.organization.agentText"                 
## [69] "synonyms.experts.organization.name"                      
## [70] "vernacularNames.name"                                    
## [71] "vernacularNames.language"                                
## [72] "vernacularNames.preferred"                               
## [73] "vernacularNames.references.titleCitation"                
## [74] "vernacularNames.references.citationDetail"               
## [75] "vernacularNames.references.uri"                          
## [76] "vernacularNames.references.author.agentText"             
## [77] "vernacularNames.references.author.fullName"              
## [78] "vernacularNames.references.author.organization.agentText"
## [79] "vernacularNames.references.author.organization.name"     
## [80] "vernacularNames.references.publicationDate"              
## [81] "vernacularNames.experts.agentText"                       
## [82] "vernacularNames.experts.fullName"                        
## [83] "vernacularNames.experts.organization.agentText"          
## [84] "vernacularNames.experts.organization.name"               
## [85] "descriptions.category"                                   
## [86] "descriptions.description"                                
## [87] "descriptions.language"                                   
## [88] "references.titleCitation"                                
## [89] "references.citationDetail"                               
## [90] "references.uri"                                          
## [91] "references.author.agentText"                             
## [92] "references.author.fullName"                              
## [93] "references.author.organization.agentText"                
## [94] "references.author.organization.name"                     
## [95] "references.publicationDate"                              
## [96] "experts.agentText"                                       
## [97] "experts.fullName"                                        
## [98] "experts.organization.agentText"                          
## [99] "experts.organization.name"</code></pre>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb72-1" data-line-number="1"><span class="co"># get info e.g. for field collectionType</span></a>
<a class="sourceLine" id="cb72-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content<span class="op">$</span>synonyms.author</a></code></pre></div>
<pre><code>## $indexed
## [1] TRUE
## 
## $type
## [1] "keyword"
## 
## $allowedOperators
##  [1] "EQUALS"             "NOT_EQUALS"         "EQUALS_IC"         
##  [4] "NOT_EQUALS_IC"      "CONTAINS"           "NOT_CONTAINS"      
##  [7] "IN"                 "NOT_IN"             "MATCHES"           
## [10] "NOT_MATCHES"        "STARTS_WITH"        "NOT_STARTS_WITH"   
## [13] "STARTS_WITH_IC"     "NOT_STARTS_WITH_IC"</code></pre>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb74-1" data-line-number="1"><span class="co"># get all settings</span></a>
<a class="sourceLine" id="cb74-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_settings</span>()</a></code></pre></div>
<pre><code>## &lt;Response&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     content: list
##     initialize: function (content, response) 
##     response: response</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb76-1" data-line-number="1"><span class="co"># get specific setting</span></a>
<a class="sourceLine" id="cb76-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" data-line-number="1"><span class="co"># check if operator is allowed</span></a>
<a class="sourceLine" id="cb78-2" data-line-number="2">tc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"synonyms.author"</span>, <span class="st">"EQUALS"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div id="available-paths-and-fields-1" class="section level4">
<h4 class="hasAnchor">
<a href="#available-paths-and-fields-1" class="anchor"></a>Available paths and fields</h4>
<p>All fields can be retrieved with <code>get_paths</code> and specific information on the fields, such as allowed operators etc. with <code>get_field_info</code></p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb80-1" data-line-number="1">tc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a>
<a class="sourceLine" id="cb80-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content</a></code></pre></div>
</div>
<div id="settings-1" class="section level4">
<h4 class="hasAnchor">
<a href="#settings-1" class="anchor"></a>Settings</h4>
<p>Multimedia-specific settings can be retrieved with <code>get_settings</code> and a specific setting with <code>get_setting</code>:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb81-1" data-line-number="1">tc<span class="op">$</span><span class="kw">get_settings</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $index.max_result_window
## [1] 50000
## 
## $taxon.group_by_scientific_name.max_num_buckets
## [1] 10000</code></pre>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb83-1" data-line-number="1">tc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
</div>
<div id="operators-1" class="section level4">
<h4 class="hasAnchor">
<a href="#operators-1" class="anchor"></a>Operators</h4>
<p>To test if a certain operator can be used for a mutimedia query:</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb85-1" data-line-number="1">tc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"identifications.defaultClassification.genus"</span>, <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## Warning in self$handleError(response): Status code:500
## Internal Server Error
## Exception: Invalid element "identifications" in path "identifications.defaultClassification.genus"
## Exception type: nl.naturalis.nba.api.NoSuchFieldException
## Full stack trace stored in response object</code></pre>
<pre><code>## [1] "Internal Server Error"</code></pre>
</div>
</div>
<div id="dwca-download-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#dwca-download-services-1" class="anchor"></a>DwCA download services</h3>
<p>The taxonomic information in the NBA is also available as Darwin-Core archive files.</p>
<div id="static-download-1" class="section level4">
<h4 class="hasAnchor">
<a href="#static-download-1" class="anchor"></a>Static download</h4>
<p>Static download services offer the download of predefined datasets. The sets that are available for download can be queried with <code>dwca_get_data_set_names</code>:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb88-1" data-line-number="1">tc<span class="op">$</span><span class="kw">dwca_get_data_set_names</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "nsr"</code></pre>
<p>A dataset can then be downloaded using <code>dwca_get_data_set</code>. A filename can be given as argument, if none is given, the DwCA archive is written to <em>download-YYYY-MM-DDThh:mm.zip</em> in the current working directory.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb90-1" data-line-number="1"><span class="co"># download dataset 'nsr' to temporary file</span></a>
<a class="sourceLine" id="cb90-2" data-line-number="2">filename &lt;-<span class="st"> </span><span class="kw">tempfile</span>(<span class="dt">fileext=</span><span class="st">".zip"</span>)</a>
<a class="sourceLine" id="cb90-3" data-line-number="3">tc<span class="op">$</span><span class="kw">dwca_get_data_set</span>(<span class="st">'nsr'</span>, <span class="dt">filename=</span>filename)                                                                               </a></code></pre></div>
<pre><code>## 
Downloading: 3.7 kB     
Downloading: 3.7 kB     
Downloading: 16 kB     
Downloading: 16 kB     
Downloading: 25 kB     
Downloading: 25 kB     
Downloading: 33 kB     
Downloading: 33 kB     
Downloading: 33 kB     
Downloading: 33 kB     
Downloading: 47 kB     
Downloading: 47 kB     
Downloading: 64 kB     
Downloading: 64 kB     
Downloading: 66 kB     
Downloading: 66 kB     
Downloading: 66 kB     
Downloading: 66 kB     
Downloading: 80 kB     
Downloading: 80 kB     
Downloading: 83 kB     
Downloading: 83 kB     
Downloading: 83 kB     
Downloading: 83 kB     
Downloading: 86 kB     
Downloading: 86 kB     
Downloading: 100 kB     
Downloading: 100 kB     
Downloading: 120 kB     
Downloading: 120 kB     
Downloading: 120 kB     
Downloading: 120 kB     
Downloading: 130 kB     
Downloading: 130 kB     
Downloading: 130 kB     
Downloading: 130 kB     
Downloading: 130 kB     
Downloading: 130 kB     
Downloading: 140 kB     
Downloading: 140 kB     
Downloading: 160 kB     
Downloading: 160 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 170 kB     
Downloading: 180 kB     
Downloading: 180 kB     
Downloading: 180 kB     
Downloading: 180 kB     
Downloading: 190 kB     
Downloading: 190 kB     
Downloading: 210 kB     
Downloading: 210 kB     
Downloading: 220 kB     
Downloading: 220 kB     
Downloading: 220 kB     
Downloading: 220 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 230 kB     
Downloading: 250 kB     
Downloading: 250 kB     
Downloading: 270 kB     
Downloading: 270 kB     
Downloading: 270 kB     
Downloading: 270 kB     
Downloading: 270 kB     
Downloading: 270 kB     
Downloading: 280 kB     
Downloading: 280 kB     
Downloading: 280 kB     
Downloading: 280 kB     
Downloading: 280 kB     
Downloading: 280 kB     
Downloading: 280 kB     
Downloading: 280 kB     
Downloading: 300 kB     
Downloading: 300 kB     
Downloading: 310 kB     
Downloading: 310 kB     
Downloading: 330 kB     
Downloading: 330 kB     
Downloading: 330 kB     
Downloading: 330 kB     
Downloading: 330 kB     
Downloading: 330 kB     
Downloading: 330 kB     
Downloading: 330 kB     
Downloading: 350 kB     
Downloading: 350 kB     
Downloading: 360 kB     
Downloading: 360 kB     
Downloading: 360 kB     
Downloading: 360 kB     
Downloading: 370 kB     
Downloading: 370 kB     
Downloading: 380 kB     
Downloading: 380 kB     
Downloading: 380 kB     
Downloading: 380 kB     
Downloading: 390 kB     
Downloading: 390 kB     
Downloading: 400 kB     
Downloading: 400 kB     
Downloading: 410 kB     
Downloading: 410 kB     
Downloading: 410 kB     
Downloading: 410 kB     
Downloading: 420 kB     
Downloading: 420 kB     
Downloading: 430 kB     
Downloading: 430 kB     
Downloading: 430 kB     
Downloading: 430 kB     
Downloading: 450 kB     
Downloading: 450 kB     
Downloading: 460 kB     
Downloading: 460 kB     
Downloading: 460 kB     
Downloading: 460 kB     
Downloading: 480 kB     
Downloading: 480 kB     
Downloading: 480 kB     
Downloading: 480 kB     
Downloading: 480 kB     
Downloading: 480 kB     
Downloading: 480 kB     
Downloading: 480 kB     
Downloading: 500 kB     
Downloading: 500 kB     
Downloading: 510 kB     
Downloading: 510 kB     
Downloading: 510 kB     
Downloading: 510 kB     
Downloading: 530 kB     
Downloading: 530 kB     
Downloading: 530 kB     
Downloading: 530 kB     
Downloading: 530 kB     
Downloading: 530 kB     
Downloading: 550 kB     
Downloading: 550 kB     
Downloading: 560 kB     
Downloading: 560 kB     
Downloading: 560 kB     
Downloading: 560 kB     
Downloading: 570 kB     
Downloading: 570 kB     
Downloading: 580 kB     
Downloading: 580 kB     
Downloading: 580 kB     
Downloading: 580 kB     
Downloading: 600 kB     
Downloading: 600 kB     
Downloading: 610 kB     
Downloading: 610 kB     
Downloading: 610 kB     
Downloading: 610 kB     
Downloading: 610 kB     
Downloading: 610 kB     
Downloading: 630 kB     
Downloading: 630 kB     
Downloading: 630 kB     
Downloading: 630 kB     
Downloading: 650 kB     
Downloading: 650 kB     
Downloading: 660 kB     
Downloading: 660 kB     
Downloading: 660 kB     
Downloading: 660 kB     
Downloading: 660 kB     
Downloading: 660 kB     
Downloading: 670 kB     
Downloading: 670 kB     
Downloading: 680 kB     
Downloading: 680 kB     
Downloading: 680 kB     
Downloading: 680 kB     
Downloading: 700 kB     
Downloading: 700 kB     
Downloading: 710 kB     
Downloading: 710 kB     
Downloading: 710 kB     
Downloading: 710 kB     
Downloading: 730 kB     
Downloading: 730 kB     
Downloading: 730 kB     
Downloading: 730 kB     
Downloading: 730 kB     
Downloading: 730 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 750 kB     
Downloading: 770 kB     
Downloading: 770 kB     
Downloading: 780 kB     
Downloading: 780 kB     
Downloading: 780 kB     
Downloading: 780 kB     
Downloading: 790 kB     
Downloading: 790 kB     
Downloading: 800 kB     
Downloading: 800 kB     
Downloading: 800 kB     
Downloading: 800 kB     
Downloading: 810 kB     
Downloading: 810 kB     
Downloading: 830 kB     
Downloading: 830 kB     
Downloading: 830 kB     
Downloading: 830 kB     
Downloading: 830 kB     
Downloading: 830 kB     
Downloading: 840 kB     
Downloading: 840 kB     
Downloading: 860 kB     
Downloading: 860 kB     
Downloading: 860 kB     
Downloading: 860 kB     
Downloading: 860 kB     
Downloading: 860 kB     
Downloading: 870 kB     
Downloading: 870 kB     
Downloading: 880 kB     
Downloading: 880 kB     
Downloading: 880 kB     
Downloading: 880 kB     
Downloading: 890 kB     
Downloading: 890 kB     
Downloading: 910 kB     
Downloading: 910 kB     
Downloading: 910 kB     
Downloading: 910 kB     
Downloading: 910 kB     
Downloading: 910 kB     
Downloading: 920 kB     
Downloading: 920 kB     
Downloading: 930 kB     
Downloading: 930 kB     
Downloading: 950 kB     
Downloading: 950 kB     
Downloading: 950 kB     
Downloading: 950 kB     
Downloading: 960 kB     
Downloading: 960 kB     
Downloading: 960 kB     
Downloading: 960 kB     
Downloading: 960 kB     
Downloading: 960 kB     
Downloading: 980 kB     
Downloading: 980 kB     
Downloading: 990 kB     
Downloading: 990 kB     
Downloading: 1,000 kB     
Downloading: 1,000 kB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.1 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.2 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.3 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.4 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.5 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.6 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.7 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.8 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 1.9 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.1 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB     
Downloading: 2.2 MB</code></pre>
<pre><code>## Query result written to /tmp/RtmpBgaqDP/file49dd7134d378.zip</code></pre>
<pre><code>## &lt;Response&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     content: NULL
##     initialize: function (content, response) 
##     response: response</code></pre>
</div>
<div id="dynamic-download-1" class="section level4">
<h4 class="hasAnchor">
<a href="#dynamic-download-1" class="anchor"></a>Dynamic download</h4>
<p>The dynamic download function <code>dwca_query</code> allows for download of arbitrary sets, defined by the user’s query. The arguments to this methods are similar to <code>query</code>, plus the filename:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb94-1" data-line-number="1"><span class="co"># download all taxa for genus 'Clematis'</span></a>
<a class="sourceLine" id="cb94-2" data-line-number="2">filename &lt;-<span class="st"> </span><span class="kw">tempfile</span>(<span class="dt">fileext=</span><span class="st">".zip"</span>)</a>
<a class="sourceLine" id="cb94-3" data-line-number="3">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"defaultClassification.genus"</span>, <span class="dt">operator=</span><span class="st">"EQUALS"</span>, <span class="dt">value=</span><span class="st">"Clematis"</span>)))</a>
<a class="sourceLine" id="cb94-4" data-line-number="4">tc<span class="op">$</span><span class="kw">dwca_query</span>(<span class="dt">querySpec=</span>qs, <span class="dt">filename=</span>filename)</a></code></pre></div>
<pre><code>## 
Downloading: 12 kB     
Downloading: 12 kB     
Downloading: 16 kB     
Downloading: 16 kB     
Downloading: 21 kB     
Downloading: 21 kB     
Downloading: 33 kB     
Downloading: 33 kB     
Downloading: 40 kB     
Downloading: 40 kB     
Downloading: 43 kB     
Downloading: 43 kB     
Downloading: 49 kB     
Downloading: 49 kB     
Downloading: 57 kB     
Downloading: 57 kB     
Downloading: 57 kB     
Downloading: 57 kB     
Downloading: 57 kB     
Downloading: 57 kB</code></pre>
<pre><code>## Query result written to /tmp/RtmpBgaqDP/file49dd3b8965fa.zip</code></pre>
<pre><code>## &lt;Response&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     content: NULL
##     initialize: function (content, response) 
##     response: response</code></pre>
</div>
</div>
</div>
<div id="geographic-data-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#geographic-data-services" class="anchor"></a>Geographic data services</h2>
<div id="query-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services-2" class="anchor"></a>Query Services</h3>
<p>The GeoArea query service allows for detailed search within the fields of a GeoArea object. As for the other data types, query parameters can be either given as a <code>list</code> or as a <code>QuerySpec</code> object. Below, we make a simple query to get the <code>GeoArea</code> object for the Netherlands:</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb98-1" data-line-number="1"><span class="co"># instantiate client for geo areas</span></a>
<a class="sourceLine" id="cb98-2" data-line-number="2">gc &lt;-<span class="st"> </span>GeoClient<span class="op">$</span><span class="kw">new</span>()</a>
<a class="sourceLine" id="cb98-3" data-line-number="3"></a>
<a class="sourceLine" id="cb98-4" data-line-number="4"><span class="co"># query for GeoArea of the Netherlands</span></a>
<a class="sourceLine" id="cb98-5" data-line-number="5">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"locality"</span>, <span class="dt">operator=</span><span class="st">"EQUALS"</span>, <span class="dt">value=</span><span class="st">"Netherlands"</span>)</a>
<a class="sourceLine" id="cb98-6" data-line-number="6">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(qc))</a>
<a class="sourceLine" id="cb98-7" data-line-number="7">res &lt;-<span class="st"> </span>gc<span class="op">$</span><span class="kw">query</span>(qs)</a>
<a class="sourceLine" id="cb98-8" data-line-number="8"></a>
<a class="sourceLine" id="cb98-9" data-line-number="9"><span class="co"># get item</span></a>
<a class="sourceLine" id="cb98-10" data-line-number="10">res<span class="op">$</span>content<span class="op">$</span>resultSet[[<span class="dv">1</span>]]<span class="op">$</span>item</a></code></pre></div>
<pre><code>## &lt;GeoArea&gt;
##   Public:
##     areaType: Country
##     clone: function (deep = FALSE) 
##     countryNL: Nederland
##     fromJSONString: function (GeoAreaJson, typeMapping = NULL) 
##     fromList: function (GeoAreaList, typeMapping = NULL) 
##     id: 1004050@GEO
##     initialize: function (sourceSystem, sourceSystemId, recordURI, id, areaType, 
##     isoCode: NLD
##     locality: Netherlands
##     recordURI: NULL
##     shape: list
##     source: World Countries
##     sourceSystem: SourceSystem, R6
##     sourceSystemId: 1004050
##     toJSONString: function (pretty = TRUE) 
##     toList: function ()</code></pre>
</div>
<div id="data-access-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services-2" class="anchor"></a>Data access services</h3>
<div id="get_geo_json_for_locality" class="section level4">
<h4 class="hasAnchor">
<a href="#get_geo_json_for_locality" class="anchor"></a>get_geo_json_for_locality</h4>
<p>This is a convenience function to directly extract the a GeoJSON object for a specific locality. <a href="https://en.wikipedia.org/wiki/GeoJSON">GeoJSON</a> is a popular format for storing geographical point- and polygon data. To e.g. extract the GeoJSON polygon representation for the Netherlands:</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb100-1" data-line-number="1">loc &lt;-<span class="st"> "Netherlands"</span></a>
<a class="sourceLine" id="cb100-2" data-line-number="2">res &lt;-<span class="st"> </span>gc<span class="op">$</span><span class="kw">get_geo_json_for_locality</span>(loc)</a></code></pre></div>
<p>Results are returned as a <code>list</code> by default, but can be easily converted to a JSON string, e.g.</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb101-1" data-line-number="1">jsonlite<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/jsonlite/topics/fromJSON">toJSON</a></span>(res<span class="op">$</span>content)</a></code></pre></div>
</div>
<div id="count-2" class="section level4">
<h4 class="hasAnchor">
<a href="#count-2" class="anchor"></a>count</h4>
<p>For any given query (with <code>QuerySpec</code> or not), returns the count of matches instead of <code>GeoArea</code> objects:</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb102-1" data-line-number="1"><span class="co"># return count of all GeoAreas</span></a>
<a class="sourceLine" id="cb102-2" data-line-number="2">res &lt;-<span class="st"> </span>gc<span class="op">$</span><span class="kw">count</span>()</a>
<a class="sourceLine" id="cb102-3" data-line-number="3">res<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] 1654</code></pre>
</div>
</div>
<div id="aggregation-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services-2" class="anchor"></a>Aggregation services</h3>
<div id="get_distinct_values-2" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values-2" class="anchor"></a>get_distinct_values</h4>
<p>This function returns all values present for a certain field, and their counts:</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb104-1" data-line-number="1">gc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"areaType"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $Nature
## [1] 987
## 
## $Municipality
## [1] 408
## 
## $Country
## [1] 247
## 
## $Province
## [1] 12</code></pre>
</div>
</div>
<div id="metadata-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services-2" class="anchor"></a>Metadata services</h3>
<p>Geo Metadata services include the same standard metadata services as for the other data types:</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb106-1" data-line-number="1"><span class="co"># get all paths for the GeoArea datatype </span></a>
<a class="sourceLine" id="cb106-2" data-line-number="2">gc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "sourceSystem.code" "sourceSystem.name" "sourceSystemId"   
##  [4] "recordURI"         "areaType"          "locality"         
##  [7] "shape"             "source"            "isoCode"          
## [10] "countryNL"</code></pre>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb108-1" data-line-number="1"><span class="co"># get info e.g. for field 'areaType'</span></a>
<a class="sourceLine" id="cb108-2" data-line-number="2">gc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content<span class="op">$</span>areaType</a></code></pre></div>
<pre><code>## $indexed
## [1] TRUE
## 
## $type
## [1] "keyword"
## 
## $allowedOperators
##  [1] "EQUALS"             "NOT_EQUALS"         "EQUALS_IC"         
##  [4] "NOT_EQUALS_IC"      "IN"                 "NOT_IN"            
##  [7] "MATCHES"            "NOT_MATCHES"        "STARTS_WITH"       
## [10] "NOT_STARTS_WITH"    "STARTS_WITH_IC"     "NOT_STARTS_WITH_IC"</code></pre>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb110-1" data-line-number="1"><span class="co"># get all settings</span></a>
<a class="sourceLine" id="cb110-2" data-line-number="2">gc<span class="op">$</span><span class="kw">get_settings</span>()</a></code></pre></div>
<pre><code>## &lt;Response&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     content: list
##     initialize: function (content, response) 
##     response: response</code></pre>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb112-1" data-line-number="1"><span class="co"># get specific setting</span></a>
<a class="sourceLine" id="cb112-2" data-line-number="2">gc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb114-1" data-line-number="1"><span class="co"># check if operator is allowed</span></a>
<a class="sourceLine" id="cb114-2" data-line-number="2">gc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"locality"</span>, <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="multimedia-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#multimedia-services" class="anchor"></a>Multimedia services</h2>
<p>Multimedia services are accessible with a <code>MultimediaClient</code>, instantiated as follows:</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb116-1" data-line-number="1">mc &lt;-<span class="st"> </span>MultimediaClient<span class="op">$</span><span class="kw">new</span>()</a></code></pre></div>
<div id="query-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services-3" class="anchor"></a>Query Services</h3>
<p>As for the other data types, the <code>query</code> method enables simple and complex queries using a list or a <code>QuerySpec</code> object to specify query parameters.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb117-1" data-line-number="1"><span class="co"># example of multimedia query passing parameters as a list</span></a>
<a class="sourceLine" id="cb117-2" data-line-number="2">mc<span class="op">$</span><span class="kw">query</span>(<span class="dt">queryParams=</span><span class="kw">list</span>(<span class="dt">collectionType=</span><span class="st">'Cnidaria'</span>))<span class="op">$</span>content</a></code></pre></div>
<pre><code>## &lt;QueryResult&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     fromJSONString: function (QueryResultJson, typeMapping = NULL) 
##     fromList: function (QueryResultList, typeMapping = NULL) 
##     initialize: function (totalSize, resultSet) 
##     resultSet: list
##     toJSONString: function (pretty = TRUE) 
##     toList: function () 
##     totalSize: 42257</code></pre>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb119-1" data-line-number="1"><span class="co"># example of multimedia query using QuerySpec: get the first 100</span></a>
<a class="sourceLine" id="cb119-2" data-line-number="2"><span class="co"># multimedia items associated with a specimen with name starting with "Ba"</span></a>
<a class="sourceLine" id="cb119-3" data-line-number="3">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"identifications.scientificName.fullScientificName"</span>, <span class="dt">operator=</span><span class="st">"STARTS_WITH"</span>, <span class="dt">value=</span><span class="st">"Qu"</span>)</a>
<a class="sourceLine" id="cb119-4" data-line-number="4">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(qc), <span class="dt">size=</span><span class="dv">100</span>)</a>
<a class="sourceLine" id="cb119-5" data-line-number="5">res &lt;-<span class="st"> </span>mc<span class="op">$</span><span class="kw">query</span>(qs)</a>
<a class="sourceLine" id="cb119-6" data-line-number="6"></a>
<a class="sourceLine" id="cb119-7" data-line-number="7"><span class="co"># check if scientific names indeed start with 'Qu'</span></a>
<a class="sourceLine" id="cb119-8" data-line-number="8"><span class="kw">sapply</span>(res<span class="op">$</span>content<span class="op">$</span>resultSet, <span class="cf">function</span>(x)x<span class="op">$</span>item<span class="op">$</span>identifications[[<span class="dv">1</span>]]<span class="op">$</span>scientificName<span class="op">$</span>fullScientificName)</a></code></pre></div>
<pre><code>##   [1] "Quercus rubra L."                             
##   [2] "Quercus rubra L."                             
##   [3] "Quercus rubra L."                             
##   [4] "Quercus rubra L."                             
##   [5] "Quercus robur L."                             
##   [6] "Quercus robur L."                             
##   [7] "Quercus robur L."                             
##   [8] "Quercus robur L."                             
##   [9] "Quercus robur L."                             
##  [10] "Quercus robur L."                             
##  [11] "Quercus robur L."                             
##  [12] "Quercus robur L."                             
##  [13] "Quercus robur L."                             
##  [14] "Quercus robur L."                             
##  [15] "Quercus robur L."                             
##  [16] "Quercus robur L."                             
##  [17] "Quercus robur L."                             
##  [18] "Quercus robur L."                             
##  [19] "Quercus robur L."                             
##  [20] "Quercus robur L."                             
##  [21] "Quercus robur L."                             
##  [22] "Quercus petraea (Matt.) Liebl."               
##  [23] "Quercus petraea (Matt.) Liebl."               
##  [24] "Quercus cerris L."                            
##  [25] "Quercus cerris L."                            
##  [26] "Quercus cerris L."                            
##  [27] "Quercus cerris L."                            
##  [28] "Quercus cerris L."                            
##  [29] "Quercus petraea (Matt.) Liebl."               
##  [30] "Quercus petraea (Matt.) Liebl."               
##  [31] "Quercus petraea (Matt.) Liebl."               
##  [32] "Quadrella cynophallophora (L.) Hutch."        
##  [33] "Quadrella cynophallophora (L.) Hutch."        
##  [34] "Quadrella cynophallophora (L.) Hutch."        
##  [35] "Quadrella cynophallophora (L.) Hutch."        
##  [36] "Quadrella cynophallophora (L.) Hutch."        
##  [37] "Quadrella cynophallophora (L.) Hutch."        
##  [38] "Quadrella indica (L.) Iltis &amp; Cornejo"        
##  [39] "Quadrella odoratissima (Jacq.) Hutch."        
##  [40] "Quadrella odoratissima (Jacq.) Hutch."        
##  [41] "Quadrella indica (L.) Iltis &amp; Cornejo"        
##  [42] "Quadrella indica (L.) Iltis &amp; Cornejo"        
##  [43] "Quadrella indica (L.) Iltis &amp; Cornejo"        
##  [44] "Quadrella indica (L.) Iltis &amp; Cornejo"        
##  [45] "Quesnelia edmundoi L.B.Sm."                   
##  [46] "Quirella  humilis (Preston, 1905)"            
##  [47] "Quadrula apiculata (Say, 1829)"               
##  [48] "Quercus gemelliflora Blume"                   
##  [49] "Quercus gemelliflora Blume"                   
##  [50] "Quercus gemelliflora Blume"                   
##  [51] "Quercus gemelliflora Blume"                   
##  [52] "Quercus gemelliflora Blume"                   
##  [53] "Quercus gemelliflora Blume"                   
##  [54] "Quercus gemelliflora Blume"                   
##  [55] "Quercus gemelliflora Blume"                   
##  [56] "Quercus gemelliflora Blume"                   
##  [57] "Quercus gemelliflora Blume"                   
##  [58] "Quercus frainetto Ten."                       
##  [59] "Quercus frainetto Ten."                       
##  [60] "Quercus frainetto Ten."                       
##  [61] "Quercus franchetii Skan"                      
##  [62] "Quercus falcata Michx."                       
##  [63] "Quercus dalechampii Ten."                     
##  [64] "Quercus argentata Korth."                     
##  [65] "Quercus argentata Korth."                     
##  [66] "Quercus argentata Korth."                     
##  [67] "Quercus argentata Korth."                     
##  [68] "Quercus argentata Korth."                     
##  [69] "Quercus argentata Korth."                     
##  [70] "Quercus aliena Blume var. acutiserrata Maxim."
##  [71] "Quercus alba L."                              
##  [72] "Quercus acutissima Carr"                      
##  [73] "Quercus acutissima Carr"                      
##  [74] "Quercus acutissima Carr"                      
##  [75] "Quercus macrocarpa Michx."                    
##  [76] "Quercus magnoliifolia Née"                    
##  [77] "Quercus magnoliifolia Née"                    
##  [78] "Quercus lyrata Walter"                        
##  [79] "Quercus lyrata Walter"                        
##  [80] "Quercus faginea Lam."                         
##  [81] "Quercus lusitanica Lam."                      
##  [82] "Quercus lusitanica Lam."                      
##  [83] "Quercus lusitanica Lam."                      
##  [84] "Quercus lusitanica Lam."                      
##  [85] "Quercus lusitanica Lam."                      
##  [86] "Quercus lusitanica Lam."                      
##  [87] "Quercus lusitanica Lam."                      
##  [88] "Quercus lowii King"                           
##  [89] "Quercus lowii King"                           
##  [90] "Quercus lineata Blume"                        
##  [91] "Quercus lineata Blume"                        
##  [92] "Quercus lineata Blume"                        
##  [93] "Quercus lineata Blume"                        
##  [94] "Quercus lineata Blume"                        
##  [95] "Quercus lanata Sm."                           
##  [96] "Quercus kerrii Craib"                         
##  [97] "Quercus coccinea Münchh."                     
##  [98] "Quercus coccinea Münchh."                     
##  [99] "Quercus coccinea Münchh."                     
## [100] "Quercus coccifera L."</code></pre>
</div>
<div id="data-access-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services-3" class="anchor"></a>Data access services</h3>
<div id="count-3" class="section level4">
<h4 class="hasAnchor">
<a href="#count-3" class="anchor"></a>count</h4>
<p>As for the other data types, a <code>count</code> function returns counts instead of the actual objects:</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb121-1" data-line-number="1"><span class="co"># count all multimedia documents</span></a>
<a class="sourceLine" id="cb121-2" data-line-number="2">mc<span class="op">$</span><span class="kw">count</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] 5364224</code></pre>
</div>
</div>
<div id="aggregation-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services-3" class="anchor"></a>Aggregation services</h3>
<div id="get_distinct_values-3" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values-3" class="anchor"></a>get_distinct_values</h4>
<p>This function returns all values present for a certain field, and their counts. Example: retrieve all different licenses and their counts"</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb123-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"license"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $CC0
## [1] 5302349
## 
## $`All rights reserved`
## [1] 33903
## 
## $`CC BY-NC-ND`
## [1] 14286
## 
## $`CC BY`
## [1] 8790
## 
## $`CC BY-NC`
## [1] 3077
## 
## $`CC BY-NC-SA`
## [1] 1255
## 
## $`CC BY-SA`
## [1] 384
## 
## $`CC BY-ND`
## [1] 2
## 
## $`CC BY-SA 3.0`
## [1] 1</code></pre>
</div>
</div>
<div id="metadata-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services-3" class="anchor"></a>Metadata services</h3>
<div id="available-paths-and-fields-2" class="section level4">
<h4 class="hasAnchor">
<a href="#available-paths-and-fields-2" class="anchor"></a>Available paths and fields</h4>
<p>All fields can be retrieved with <code>get_paths</code> and specific information on the fields, such as allowed operators etc. with <code>get_field_info</code></p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb125-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a>
<a class="sourceLine" id="cb125-2" data-line-number="2">mc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content</a></code></pre></div>
</div>
<div id="settings-2" class="section level4">
<h4 class="hasAnchor">
<a href="#settings-2" class="anchor"></a>Settings</h4>
<p>Multimedia-specific settings can be retrieved with <code>get_settings</code> and a specific setting with <code>get_setting</code>:</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb126-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_settings</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $index.max_result_window
## [1] 50000</code></pre>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb128-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
</div>
<div id="operators-2" class="section level4">
<h4 class="hasAnchor">
<a href="#operators-2" class="anchor"></a>Operators</h4>
<p>To test if a certain operator can be used for a mutimedia query:</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb130-1" data-line-number="1">mc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"identifications.scientificName.fullScientificName"</span>, <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
</div>
<div id="metadata-services-4" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#metadata-services-4" class="anchor"></a>Metadata services</h2>
<p>Metadata services provide miscellaneous information about the data available via the NBA. Note that there is also type-specific metadata for each data type (e.g. <code>Specimen</code>) which can be retrieved with the specific client of that class. Here we show the available methods for the <code>MetadataClient</code> which gives general, non-type specific metadata. The client is instantiated in the standard way:</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb132-1" data-line-number="1">mc &lt;-<span class="st"> </span>MetadataClient<span class="op">$</span><span class="kw">new</span>()</a></code></pre></div>
<div id="controlled-vocabularies" class="section level4">
<h4 class="hasAnchor">
<a href="#controlled-vocabularies" class="anchor"></a>Controlled vocabularies</h4>
<p>The vocabularies for some fields are controlled by dictionaries with allowed values. For the sex of a museum specimen, for instance, only the terms <em>male</em>, <em>female</em>, <em>mixed</em> and <em>hermaphrodite</em> are allowed to be assigned to the specimen. The fields for which controlled lists are available can be retrieved as follows:</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb133-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_lists</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "PhaseOrStage"       "Sex"                "SpecimenTypeStatus"
## [4] "TaxonomicStatus"</code></pre>
<p>and for each field that has a controlled vocabulary, there is a separate function to retrieve the allowed values:</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb135-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_list_taxonomic_status</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "accepted name"               "synonym"                    
## [3] "basionym"                    "homonym"                    
## [5] "ambiguous synonym"           "misapplied name"            
## [7] "misspelled name"             "provisionally accepted name"</code></pre>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb137-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_list_specimen_type_status</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "allotype"      "epitype"       "hapantotype"   "holotype"     
##  [5] "isoepitype"    "isolectotype"  "isoneotype"    "isosyntype"   
##  [9] "isotype"       "lectotype"     "neotype"       "paratype"     
## [13] "paralectotype" "syntype"       "topotype"      "type"</code></pre>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb139-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_list_sex</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "male"          "female"        "mixed"         "hermaphrodite"</code></pre>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb141-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_list_phase_or_stage</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "adult"    "subadult" "egg"      "embryo"   "immature" "juvenile"
## [7] "larva"    "pupa"     "nymph"</code></pre>
</div>
<div id="miscellaneous" class="section level4">
<h4 class="hasAnchor">
<a href="#miscellaneous" class="anchor"></a>Miscellaneous</h4>
<div id="dates" class="section level5">
<h5 class="hasAnchor">
<a href="#dates" class="anchor"></a>Dates</h5>
<p>To maintain data integrity, dates have to be coded in specific formats in our systems. For instance, <em>yyyy-MM-dd</em> is a valid format. Allowed formats can be retrieved as follows:</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb143-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_allowed_date_formats</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "yyyy-MM-dd'T'HH:mm:ssZ"         "yyyy-MM-dd'T'HH:mm[:ss][.SSS]Z"
## [3] "yyyy-MM-dd'T'HH:mm[:ss]"        "yyyy-MM-dd HH:mm[:ss]"         
## [5] "yyyy-MM-dd"                     "yyyy-MM"                       
## [7] "yyyy"</code></pre>
</div>
<div id="services-list" class="section level5">
<h5 class="hasAnchor">
<a href="#services-list" class="anchor"></a>Services list</h5>
<p>The method <code>get_rest_services</code> returns a list of all services in the NBA as objects of type <code>RestService</code>.</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb145-1" data-line-number="1"><span class="co"># get the endPoint of the first rest service in the services list</span></a>
<a class="sourceLine" id="cb145-2" data-line-number="2">mc<span class="op">$</span><span class="kw">get_rest_services</span>()<span class="op">$</span>content[[<span class="dv">1</span>]]</a></code></pre></div>
<pre><code>## &lt;RestService&gt;
##   Public:
##     clone: function (deep = FALSE) 
##     consumes: application/json;charset=UTF-8
##     endPoint: /taxon/query
##     fromJSONString: function (RestServiceJson, typeMapping = NULL) 
##     fromList: function (RestServiceList, typeMapping = NULL) 
##     initialize: function (endPoint, method, consumes, produces, url) 
##     method: POST
##     produces: application/json;charset=UTF-8
##     toJSONString: function (pretty = TRUE) 
##     toList: function () 
##     url: http://api.biodiversitydata.nl/v2/taxon/query</code></pre>
</div>
<div id="settings-3" class="section level5">
<h5 class="hasAnchor">
<a href="#settings-3" class="anchor"></a>Settings</h5>
<p>Similar to the document-specific metadata services, we can get general settings with the <code>MetaDataClient</code>:</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb147-1" data-line-number="1"><span class="co">#get all settings</span></a>
<a class="sourceLine" id="cb147-2" data-line-number="2">mc<span class="op">$</span><span class="kw">get_settings</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $operator.contains.min_term_length
## [1] 3
## 
## $operator.contains.max_term_length
## [1] 15</code></pre>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb149-1" data-line-number="1"><span class="co"># get value for specific setting</span></a>
<a class="sourceLine" id="cb149-2" data-line-number="2">mc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"operator.contains.min_term_length"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 3</code></pre>
</div>
</div>
</div>
</div>
<div id="use-cases" class="section level1">
<h1 class="hasAnchor">
<a href="#use-cases" class="anchor"></a>Use cases</h1>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb151-1" data-line-number="1">## all records with a chronoIdentifier</a>
<a class="sourceLine" id="cb151-2" data-line-number="2">sc &lt;-<span class="st"> </span>SpecimenClient<span class="op">$</span><span class="kw">new</span>()</a>
<a class="sourceLine" id="cb151-3" data-line-number="3"></a>
<a class="sourceLine" id="cb151-4" data-line-number="4">## get entries of family "Carcharhinidae" that have youngChronoName or oldChronoName</a>
<a class="sourceLine" id="cb151-5" data-line-number="5">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"gatheringEvent.chronoStratigraphy.youngChronoName"</span>, <span class="dt">operator=</span><span class="st">"NOT_EQUALS"</span>, <span class="dt">value=</span><span class="st">""</span>)</a>
<a class="sourceLine" id="cb151-6" data-line-number="6">qc2 &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"gatheringEvent.chronoStratigraphy.oldChronoName"</span>, <span class="dt">operator=</span><span class="st">"NOT_EQUALS"</span>, <span class="dt">value=</span><span class="st">""</span>)</a>
<a class="sourceLine" id="cb151-7" data-line-number="7">qc<span class="op">$</span>or &lt;-<span class="st"> </span><span class="kw">list</span>(qc2)</a>
<a class="sourceLine" id="cb151-8" data-line-number="8">shark_orders &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Carcharhiniformes"</span>,</a>
<a class="sourceLine" id="cb151-9" data-line-number="9">                  <span class="st">"Heterodontiformes"</span>,</a>
<a class="sourceLine" id="cb151-10" data-line-number="10">                  <span class="st">"Hexanchiformes"</span>,</a>
<a class="sourceLine" id="cb151-11" data-line-number="11">                  <span class="st">"Lamniformes"</span>,</a>
<a class="sourceLine" id="cb151-12" data-line-number="12">                  <span class="st">"Orectolobiformes"</span>,</a>
<a class="sourceLine" id="cb151-13" data-line-number="13">                  <span class="st">"Pristiophoriformes"</span>,</a>
<a class="sourceLine" id="cb151-14" data-line-number="14">                  <span class="st">"Squaliformes"</span>,</a>
<a class="sourceLine" id="cb151-15" data-line-number="15">                  <span class="st">"Squatiniformes"</span>)</a>
<a class="sourceLine" id="cb151-16" data-line-number="16">qc3 &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">"identifications.defaultClassification.order"</span>, <span class="dt">operator=</span><span class="st">"IN"</span>, <span class="dt">value=</span>shark_orders) </a>
<a class="sourceLine" id="cb151-17" data-line-number="17"></a>
<a class="sourceLine" id="cb151-18" data-line-number="18"></a>
<a class="sourceLine" id="cb151-19" data-line-number="19">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw">list</span>(qc, qc3), <span class="dt">size=</span><span class="dv">5000</span>)</a>
<a class="sourceLine" id="cb151-20" data-line-number="20">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">query</span>(<span class="dt">querySpec =</span> qs)</a>
<a class="sourceLine" id="cb151-21" data-line-number="21"></a>
<a class="sourceLine" id="cb151-22" data-line-number="22">## get geological periods: youngChronoName and oldChrononame</a>
<a class="sourceLine" id="cb151-23" data-line-number="23">genera &lt;-<span class="st"> </span><span class="kw">sapply</span>(res<span class="op">$</span>content<span class="op">$</span>resultSet, <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb151-24" data-line-number="24">    res &lt;-<span class="st"> </span>x<span class="op">$</span>item<span class="op">$</span>identifications[[<span class="dv">1</span>]]<span class="op">$</span>defaultClassification<span class="op">$</span>genus</a>
<a class="sourceLine" id="cb151-25" data-line-number="25">    <span class="kw">ifelse</span>(<span class="kw">is.null</span>(res), <span class="ot">NA</span>, res)</a>
<a class="sourceLine" id="cb151-26" data-line-number="26">})</a>
<a class="sourceLine" id="cb151-27" data-line-number="27"></a>
<a class="sourceLine" id="cb151-28" data-line-number="28">species &lt;-<span class="st"> </span><span class="kw">sapply</span>(res<span class="op">$</span>content<span class="op">$</span>resultSet, <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb151-29" data-line-number="29">    res &lt;-<span class="st"> </span>x<span class="op">$</span>item<span class="op">$</span>identifications[[<span class="dv">1</span>]]<span class="op">$</span>defaultClassification<span class="op">$</span>specificEpithet</a>
<a class="sourceLine" id="cb151-30" data-line-number="30">    <span class="kw">ifelse</span>(<span class="kw">is.null</span>(res), <span class="ot">NA</span>, res)</a>
<a class="sourceLine" id="cb151-31" data-line-number="31">})</a>
<a class="sourceLine" id="cb151-32" data-line-number="32"></a>
<a class="sourceLine" id="cb151-33" data-line-number="33">young_periods &lt;-<span class="st"> </span><span class="kw">sapply</span>(res<span class="op">$</span>content<span class="op">$</span>resultSet, <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb151-34" data-line-number="34">    res &lt;-<span class="st"> </span>x<span class="op">$</span>item<span class="op">$</span>gatheringEvent<span class="op">$</span>chronoStratigraphy[[<span class="dv">1</span>]]<span class="op">$</span>youngChronoName</a>
<a class="sourceLine" id="cb151-35" data-line-number="35">    <span class="kw">ifelse</span>(<span class="kw">is.null</span>(res), <span class="ot">NA</span>, res)</a>
<a class="sourceLine" id="cb151-36" data-line-number="36">})</a>
<a class="sourceLine" id="cb151-37" data-line-number="37"></a>
<a class="sourceLine" id="cb151-38" data-line-number="38">old_periods &lt;-<span class="st"> </span><span class="kw">sapply</span>(res<span class="op">$</span>content<span class="op">$</span>resultSet, <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb151-39" data-line-number="39">    res &lt;-<span class="st"> </span>x<span class="op">$</span>item<span class="op">$</span>gatheringEvent<span class="op">$</span>chronoStratigraphy[[<span class="dv">1</span>]]<span class="op">$</span>oldChronoName</a>
<a class="sourceLine" id="cb151-40" data-line-number="40">    <span class="kw">ifelse</span>(<span class="kw">is.null</span>(res), <span class="ot">NA</span>, res)</a>
<a class="sourceLine" id="cb151-41" data-line-number="41">})</a>
<a class="sourceLine" id="cb151-42" data-line-number="42"></a>
<a class="sourceLine" id="cb151-43" data-line-number="43">## translate the geological periods to mya</a>
<a class="sourceLine" id="cb151-44" data-line-number="44">times &lt;-<span class="st"> </span><span class="kw"><a href="../reference/geo_age.html">geo_age</a></span>(<span class="kw">unique</span>(<span class="kw">c</span>(young_periods, old_periods)))</a>
<a class="sourceLine" id="cb151-45" data-line-number="45"></a>
<a class="sourceLine" id="cb151-46" data-line-number="46">## put all relevant data together</a>
<a class="sourceLine" id="cb151-47" data-line-number="47">data &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">genus=</span>genera, <span class="dt">species=</span>species, <span class="dt">young_p=</span>young_periods, <span class="dt">old_p=</span>old_periods)</a>
<a class="sourceLine" id="cb151-48" data-line-number="48">data<span class="op">$</span>young_age &lt;-<span class="st"> </span><span class="kw">sapply</span>(data<span class="op">$</span>young_p, <span class="cf">function</span>(x)<span class="kw">ifelse</span>(<span class="kw">is.na</span>(x), <span class="ot">NA</span>, <span class="kw">unlist</span>(times[<span class="st">'late_age'</span>, <span class="kw">as.character</span>(x)])))</a>
<a class="sourceLine" id="cb151-49" data-line-number="49">data<span class="op">$</span>old_age &lt;-<span class="st"> </span><span class="kw">sapply</span>(data<span class="op">$</span>old_p, <span class="cf">function</span>(x)<span class="kw">ifelse</span>(<span class="kw">is.na</span>(x), <span class="ot">NA</span>, <span class="kw">unlist</span>(times[<span class="st">'early_age'</span>, <span class="kw">as.character</span>(x)])))</a>
<a class="sourceLine" id="cb151-50" data-line-number="50">## filter for the ones that have data</a>
<a class="sourceLine" id="cb151-51" data-line-number="51">data &lt;-<span class="st"> </span>data[<span class="op">!</span><span class="st"> </span>(<span class="kw">is.na</span>(data<span class="op">$</span>young_age) <span class="op">&amp;</span><span class="st"> </span><span class="kw">is.na</span>(data<span class="op">$</span>old_age)), ]</a>
<a class="sourceLine" id="cb151-52" data-line-number="52"></a>
<a class="sourceLine" id="cb151-53" data-line-number="53">## filter out duplicates</a>
<a class="sourceLine" id="cb151-54" data-line-number="54">data &lt;-<span class="st"> </span><span class="kw">unique</span>(data)</a>
<a class="sourceLine" id="cb151-55" data-line-number="55"></a>
<a class="sourceLine" id="cb151-56" data-line-number="56">## set ages if one value is NA</a>
<a class="sourceLine" id="cb151-57" data-line-number="57">##data$young_age[is.na(data$young_age)] &lt;- data$old_age[is.na(data$young_age)]</a>
<a class="sourceLine" id="cb151-58" data-line-number="58">##data$old_age[is.na(data$old_age)] &lt;- data$young_age[is.na(data$old_age)]</a>
<a class="sourceLine" id="cb151-59" data-line-number="59"></a>
<a class="sourceLine" id="cb151-60" data-line-number="60"><span class="kw">require</span>(ape)</a>
<a class="sourceLine" id="cb151-61" data-line-number="61">tree &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/read.nexus">read.nexus</a></span>(<span class="st">'other/sharks.nex'</span>)</a>
<a class="sourceLine" id="cb151-62" data-line-number="62"></a>
<a class="sourceLine" id="cb151-63" data-line-number="63">tree_genera &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">sapply</span>(<span class="kw">strsplit</span>(tree<span class="op">$</span>tip.label, <span class="st">"_"</span>), <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">1</span>))</a>
<a class="sourceLine" id="cb151-64" data-line-number="64"></a>
<a class="sourceLine" id="cb151-65" data-line-number="65">## filter out genera that are not in our tree</a>
<a class="sourceLine" id="cb151-66" data-line-number="66">data_filtered &lt;-<span class="st"> </span>data[data<span class="op">$</span>genus <span class="op">%in%</span><span class="st"> </span>tree_genera,]</a>
<a class="sourceLine" id="cb151-67" data-line-number="67"></a>
<a class="sourceLine" id="cb151-68" data-line-number="68">genera_ages &lt;-<span class="st"> </span><span class="kw">aggregate</span>(data_filtered[<span class="kw">c</span>(<span class="st">'young_age'</span>, <span class="st">'old_age'</span>)], <span class="dt">by=</span><span class="kw">list</span>(data_filtered<span class="op">$</span>genus), <span class="dt">FUN=</span>mean, <span class="dt">na.rm=</span>T)</a>
<a class="sourceLine" id="cb151-69" data-line-number="69">genera_ages &lt;-<span class="st"> </span>genera_ages[<span class="kw">complete.cases</span>(genera_ages),]</a>
<a class="sourceLine" id="cb151-70" data-line-number="70"></a>
<a class="sourceLine" id="cb151-71" data-line-number="71">genera_ages<span class="op">$</span>node &lt;-<span class="st"> </span><span class="kw">sapply</span>(genera_ages[,<span class="dv">1</span>], <span class="cf">function</span>(genus) {</a>
<a class="sourceLine" id="cb151-72" data-line-number="72">    tips &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="kw">paste0</span>(<span class="st">"^"</span>, genus, <span class="st">"_"</span>), tree<span class="op">$</span>tip.label)</a>
<a class="sourceLine" id="cb151-73" data-line-number="73">    mrca &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/mrca">getMRCA</a></span>(tree, tips)</a>
<a class="sourceLine" id="cb151-74" data-line-number="74">    <span class="kw">ifelse</span>(<span class="kw">is.null</span>(mrca), <span class="ot">NA</span>, mrca)</a>
<a class="sourceLine" id="cb151-75" data-line-number="75">})</a>
<a class="sourceLine" id="cb151-76" data-line-number="76"></a>
<a class="sourceLine" id="cb151-77" data-line-number="77">genera_ages &lt;-<span class="st"> </span>genera_ages[<span class="kw">complete.cases</span>(genera_ages),]</a>
<a class="sourceLine" id="cb151-78" data-line-number="78"></a>
<a class="sourceLine" id="cb151-79" data-line-number="79">## We have a rogue taxon in Genus Squatina. Filter it out!!</a>
<a class="sourceLine" id="cb151-80" data-line-number="80">genera_ages &lt;-<span class="st"> </span>genera_ages[<span class="op">!</span><span class="st"> </span>genera_ages[,<span class="dv">1</span>] <span class="op">==</span><span class="st"> "Squatina"</span>,]</a>
<a class="sourceLine" id="cb151-81" data-line-number="81"><span class="co">#genera_ages &lt;- genera_ages[! genera_ages[,1] == "Carcharadon",] ## same node??</span></a>
<a class="sourceLine" id="cb151-82" data-line-number="82"><span class="co">#genera_ages &lt;- genera_ages[! genera_ages[,1] == "Carcharias",]</span></a>
<a class="sourceLine" id="cb151-83" data-line-number="83"><span class="co">#genera_ages &lt;- genera_ages[! genera_ages[,1] == "Odontaspis",] ## same node??</span></a>
<a class="sourceLine" id="cb151-84" data-line-number="84"><span class="co"># "Pristiophorus" : non-monophyletic</span></a>
<a class="sourceLine" id="cb151-85" data-line-number="85"></a>
<a class="sourceLine" id="cb151-86" data-line-number="86"></a>
<a class="sourceLine" id="cb151-87" data-line-number="87">calib &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/chronos">makeChronosCalib</a></span>(tree, genera_ages<span class="op">$</span>node, <span class="dt">age.min=</span>genera_ages<span class="op">$</span>young_age, <span class="dt">age.max=</span>genera_ages<span class="op">$</span>old_age)</a>
<a class="sourceLine" id="cb151-88" data-line-number="88">chronogram &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/chronos">chronos</a></span>(tree, <span class="dt">calibration=</span>calib)</a>
<a class="sourceLine" id="cb151-89" data-line-number="89"></a>
<a class="sourceLine" id="cb151-90" data-line-number="90"></a>
<a class="sourceLine" id="cb151-91" data-line-number="91">## get mrca</a>
<a class="sourceLine" id="cb151-92" data-line-number="92"></a>
<a class="sourceLine" id="cb151-93" data-line-number="93">##t &lt;- t(sapply(unique(data_filtered$genus), function(x)colMeans(data_filtered[data_filtered$genus==x, c('young_age', 'old_age')])))</a>
<a class="sourceLine" id="cb151-94" data-line-number="94"></a>
<a class="sourceLine" id="cb151-95" data-line-number="95">## filter out NaN values (genera that are not in our data)</a>
<a class="sourceLine" id="cb151-96" data-line-number="96">##t &lt;- dat[complete.cases(dat),]</a>
<a class="sourceLine" id="cb151-97" data-line-number="97"></a>
<a class="sourceLine" id="cb151-98" data-line-number="98"></a>
<a class="sourceLine" id="cb151-99" data-line-number="99"></a>
<a class="sourceLine" id="cb151-100" data-line-number="100">## filter data further such that we only keep the taxa that are in</a>
<a class="sourceLine" id="cb151-101" data-line-number="101">## the tree</a>
<a class="sourceLine" id="cb151-102" data-line-number="102"></a>
<a class="sourceLine" id="cb151-103" data-line-number="103">## data_filtered &lt;- data[paste(data$genus, data$species, sep="_") %in% tree$tip.label,]</a>
<a class="sourceLine" id="cb151-104" data-line-number="104">##species &lt;- paste(data$genus, data$species, sep="_")</a>
<a class="sourceLine" id="cb151-105" data-line-number="105"></a>
<a class="sourceLine" id="cb151-106" data-line-number="106">## aggregate over genus</a>
<a class="sourceLine" id="cb151-107" data-line-number="107">## aggregate(young_age~genus, data_filtered, mean)</a>
<a class="sourceLine" id="cb151-108" data-line-number="108"></a>
<a class="sourceLine" id="cb151-109" data-line-number="109">## aggregate(data, list(data_filtered$genus), function(x)ifelse(all(is.na(x)), NA, mean(x, na.rm=T)))</a>
<a class="sourceLine" id="cb151-110" data-line-number="110"></a>
<a class="sourceLine" id="cb151-111" data-line-number="111"></a>
<a class="sourceLine" id="cb151-112" data-line-number="112"></a>
<a class="sourceLine" id="cb151-113" data-line-number="113"></a>
<a class="sourceLine" id="cb151-114" data-line-number="114"></a>
<a class="sourceLine" id="cb151-115" data-line-number="115">## aggregate(.~genus, data, mean)</a>
<a class="sourceLine" id="cb151-116" data-line-number="116"></a>
<a class="sourceLine" id="cb151-117" data-line-number="117">##aggregate(young_age+old_age~genus, data, mean)</a>
<a class="sourceLine" id="cb151-118" data-line-number="118"></a>
<a class="sourceLine" id="cb151-119" data-line-number="119">##aggregate(young_age~genus, data, mean)</a>
<a class="sourceLine" id="cb151-120" data-line-number="120">##aggregate(old_age~genus, data, mean)</a>
<a class="sourceLine" id="cb151-121" data-line-number="121"></a>
<a class="sourceLine" id="cb151-122" data-line-number="122">##aggregate(young_age+old_age~genus, data, mean)</a>
<a class="sourceLine" id="cb151-123" data-line-number="123"></a>
<a class="sourceLine" id="cb151-124" data-line-number="124"></a>
<a class="sourceLine" id="cb151-125" data-line-number="125"></a>
<a class="sourceLine" id="cb151-126" data-line-number="126"></a>
<a class="sourceLine" id="cb151-127" data-line-number="127">df &lt;-<span class="st"> </span><span class="kw">aggregate</span>(data_filtered, <span class="kw">list</span>(data_filtered<span class="op">$</span>genus, data_filtered<span class="op">$</span>species), <span class="cf">function</span>(x)<span class="kw">ifelse</span>(<span class="kw">all</span>(<span class="kw">is.na</span>(x)), <span class="ot">NA</span>, <span class="kw">mean</span>(x, <span class="dt">na.rm=</span>T)))</a>
<a class="sourceLine" id="cb151-128" data-line-number="128">df<span class="op">$</span>young_age[<span class="kw">is.na</span>(df<span class="op">$</span>young_age)] &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb151-129" data-line-number="129">nodes &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">paste</span>(df<span class="op">$</span>Group<span class="fl">.1</span>, df<span class="op">$</span>Group<span class="fl">.2</span>, <span class="dt">sep=</span><span class="st">"_"</span>), <span class="cf">function</span>(x)<span class="kw">which</span>(x<span class="op">==</span>tree<span class="op">$</span>tip.label))</a>
<a class="sourceLine" id="cb151-130" data-line-number="130">nn &lt;-<span class="st"> </span><span class="kw">sapply</span>(nodes, <span class="cf">function</span>(x)<span class="kw">get_parent</span>(tree, x))</a>
<a class="sourceLine" id="cb151-131" data-line-number="131"></a>
<a class="sourceLine" id="cb151-132" data-line-number="132">calib &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/chronos">makeChronosCalib</a></span>(tree, <span class="dt">node=</span>nn, <span class="dt">age.min=</span>df<span class="op">$</span>young_age, <span class="dt">age.max=</span>df<span class="op">$</span>old_age)</a>
<a class="sourceLine" id="cb151-133" data-line-number="133">chronogram &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/chronos">chronos</a></span>(tree, <span class="dt">calibration=</span>calib)</a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#background">Background</a><ul class="nav nav-pills nav-stacked">
<li><a href="#the-netherlands-biodiversity-api">The Netherlands Biodiversity API</a></li>
      <li><a href="#r-access">R access</a></li>
      </ul>
</li>
      <li>
<a href="#getting-started">Getting started</a><ul class="nav nav-pills nav-stacked">
<li><a href="#main-data-types-and-classes">Main data types and classes</a></li>
      <li><a href="#api-client-classes">API Client classes</a></li>
      <li><a href="#queries">Queries</a></li>
      </ul>
</li>
      <li>
<a href="#advanced-queries">Advanced queries</a><ul class="nav nav-pills nav-stacked">
<li><a href="#using-the-queryspec-object">Using the QuerySpec object</a></li>
      </ul>
</li>
      <li>
<a href="#services-examples">Services examples</a><ul class="nav nav-pills nav-stacked">
<li><a href="#specimen-occurrence-services">Specimen occurrence services</a></li>
      <li><a href="#taxonomic-data-services">Taxonomic data services</a></li>
      <li><a href="#geographic-data-services">Geographic data services</a></li>
      <li><a href="#multimedia-services">Multimedia services</a></li>
      <li><a href="#metadata-services-4">Metadata services</a></li>
      </ul>
</li>
      <li><a href="#use-cases">Use cases</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hannes Hettling, Rutger Vos.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
