<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>nbaR services examples â€¢ nbaR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="nbaR services examples">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nbaR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/nbaR.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/apiclient.html">Objects and advanced queries</a>
    </li>
    <li>
      <a href="../articles/services.html">Services summary</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Example Workflows
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/sharks.html">Calibrating a molecular phylogeny of the sharks</a>
    </li>
    <li>
      <a href="../articles/tomato.html">The oldest tomato specimen in the world</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/apiclient.html">nbaR API client: concepts and object model</a>
    </li>
    <li>
      <a href="../articles/services.html">nbaR services examples</a>
    </li>
    <li>
      <a href="../articles/sharks.html">Calibrating a molecular phylogeny of the sharks</a>
    </li>
    <li>
      <a href="../articles/tomato.html">The oldest tomato specimen in the world</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/naturalis/nbaR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>nbaR services examples</h1>
            
            <h4 class="date">Apr 16, 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/naturalis/nbaR/blob/master/vignettes/services.Rmd"><code>vignettes/services.Rmd</code></a></small>
      <div class="hidden name"><code>services.Rmd</code></div>

    </div>

    
    
<div id="services-examples" class="section level1">
<h1 class="hasAnchor">
<a href="#services-examples" class="anchor"></a>Services examples</h1>
<div id="specimen-occurrence-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#specimen-occurrence-services" class="anchor"></a>Specimen occurrence services</h2>
<p>Specimen records constitute the core of data served by the NBA. Museum specimens can represent a whole variety of different objects such as plants, animals or single parts thereof, DNA samples, fossils, rocks or meteorites. For detailed information of the data model, please refer to the <a href="http://docs.biodiversitydata.nl/en/latest/doc-spec-services/specimen/">official documentation in the NBA</a>.</p>
<p>All specimen occurrence services are accessible using the methods within the <code>SpecimenClient</code> class. For a list of available endpoints, please refer to the class documentation (<code><a href="../reference/Specimen.html">?Specimen</a></code>). Below, we will give details about all services, grouped by category.</p>
<div id="query-services" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services" class="anchor"></a>Query services</h3>
<div id="query" class="section level4">
<h4 class="hasAnchor">
<a href="#query" class="anchor"></a>query</h4>
<p>Querying for specimens is accomplished with the <code>query</code> method in the class <code>SpecimenClient</code>. For simple queries, query parameters of type <code>list</code> can be passed via the parameter <code>queryParams</code>, for example we can query specimens of the Family <em>Ebenaceae</em> that were collected in Europe:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">'nbaR'</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co"># instantiate specimen client</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">sc &lt;-<span class="st"> </span>SpecimenClient<span class="op">$</span><span class="kw">new</span>()</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co"># specify query params in named list</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">qp &lt;-</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</a>
<a class="sourceLine" id="cb1-9" data-line-number="9">    <span class="dt">identifications.defaultClassification.family =</span> <span class="st">"Ebenaceae"</span>,</a>
<a class="sourceLine" id="cb1-10" data-line-number="10">    <span class="dt">gatheringEvent.continent =</span> <span class="st">"Europe"</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11">  )</a>
<a class="sourceLine" id="cb1-12" data-line-number="12"></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="co"># query</span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">query</span>(<span class="dt">queryParams =</span> qp)</a></code></pre></div>
<p>If we now want to know all the countries that the specimens were collected in, we can access the <code>Specimen</code> objects in <code>res$content$resultSet</code> as follows:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(res<span class="op">$</span>content<span class="op">$</span>resultSet, <span class="cf">function</span>(x)x<span class="op">$</span>item<span class="op">$</span>gatheringEvent<span class="op">$</span>country)</a></code></pre></div>
<pre><code>##  [1] "Austria"     "Italy"       "Netherlands" "Italy"       "Switzerland"
##  [6] "Netherlands" "Switzerland" "Netherlands" "France"      "Austria"</code></pre>
<p>Note that passing query parameters as a named list only allows for limited queries; the logical conjunction between parameters is for example always <code>AND</code>. More complex queries can be accomplished using the <code>QuerySpec</code> object:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># get all specimens with genus name starting with 'Hydro'</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">qc &lt;-</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="st">  </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span> <span class="st">"identifications.defaultClassification.genus"</span>,</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">                     <span class="dt">operator =</span> <span class="st">"STARTS_WITH"</span>,</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">                     <span class="dt">value =</span> <span class="st">"Hydro"</span>)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(qc))</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">query</span>(qs)</a></code></pre></div>
</div>
<div id="download_query" class="section level4">
<h4 class="hasAnchor">
<a href="#download_query" class="anchor"></a>download_query</h4>
<p>The <code>query</code> function is limited to retrieve 50000 specimen at once (this is determined in the parameter <code>index.max_result_window</code>, the value is retrievable using the <code>getSettings</code> method in the metadata section). In order to provide access for a larger amount of data, the <code>query_download</code> takes the same arguments as <code>query</code>, but download the data as a gzip stream under the hood. Unlike <code>query</code>, <code>query_download</code> returns a list of specimen objects instead of a <code>ResultSet</code>. Example:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">## get the first 100000 specimen objects (not possible with query method)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">download_query</span>(QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">size=</span><span class="dv">100000</span>))</a></code></pre></div>
</div>
</div>
<div id="data-access-services" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services" class="anchor"></a>Data access services</h3>
<p>Several access methods offer the convenient retrieval of specimens matching a certain identifier or being part of a certain collection. Below we give examples of how to use the currently implemented data access services for specimen records:</p>
<div id="access-to-collections" class="section level4">
<h4 class="hasAnchor">
<a href="#access-to-collections" class="anchor"></a>Access to collections</h4>
<p>Some of the specimens available via the NBA are categorised thematically into special collections, such as the Siebold-, Dubois- or Jongmans collection. The function <code>get_named_collection</code> lists all available special collections and the identifiers of the specimens within a collection can be queries with <code>get_ids_in_collection</code></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_named_collections</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "jongmans"      "dubois"        "timor"         "mammoth"      
##  [5] "siebold"       "martin"        "arts"          "meteorites"   
##  [9] "en_tibi"       "extinct-birds"</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_ids_in_collection</span>(<span class="st">"siebold"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "RMNH.CRUS.D.406@CRS"  "RMNH.CRUS.D.738@CRS"  "RMNH.CRUS.D.822@CRS" 
##  [4] "RMNH.CRUS.D.921@CRS"  "RMNH.CRUS.S.478@CRS"  "RGM.329283@CRS"      
##  [7] "RMNH.CRUS.D.1132@CRS" "RMNH.PISC.D.167@CRS"  "RMNH.RENA.1701@CRS"  
## [10] "RMNH.RENA.18559@CRS"</code></pre>
</div>
<div id="count" class="section level4">
<h4 class="hasAnchor">
<a href="#count" class="anchor"></a>count</h4>
<p>For any given query (with <code>QuerySpec</code> or not), returns the count of matches instead of specimen objects:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># Example with QuerySpec:</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="co"># how many specimens are there in the 'Botany' collection?</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3">qc &lt;-<span class="st"> </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field=</span><span class="st">'collectionType'</span>, <span class="dt">operator=</span><span class="st">'EQUALS'</span>, <span class="dt">value=</span><span class="st">'Botany'</span>)</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(qc))</a>
<a class="sourceLine" id="cb10-5" data-line-number="5"></a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="co"># get the number of specimens</span></a>
<a class="sourceLine" id="cb10-7" data-line-number="7">sc<span class="op">$</span><span class="kw">count</span>(qs)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] 4897385</code></pre>
<p>Note that the count of matches for a given query is also returned by the <code>query</code> function. However, <code>count</code> is more lightweight as it returns an <code>integer</code> instead of a <code>ResultSet</code> containing <code>Specimen</code> objects.</p>
</div>
<div id="exists" class="section level4">
<h4 class="hasAnchor">
<a href="#exists" class="anchor"></a>exists</h4>
<p>Check if a record exists, based on its <em>unitID</em>:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># use SpecimenClient instantiated above</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/exists">exists</a></span>(<span class="st">'ZMA.INS.1255440'</span>)                                        </a>
<a class="sourceLine" id="cb12-3" data-line-number="3"></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="co"># content is boolean</span></a>
<a class="sourceLine" id="cb12-5" data-line-number="5">res<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="find" class="section level4">
<h4 class="hasAnchor">
<a href="#find" class="anchor"></a>find</h4>
<p>Return a single specimen given its identifier (Note: the identifier of a specimen is different from the unitID, see also <a href="http://docs.biodiversitydata.nl/en/latest/doc-spec-services/specimen/#specids">here</a>):</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">id &lt;-<span class="st"> "RMNH.MAM.17209.B@CRS"</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/apropos">find</a></span>(id)</a>
<a class="sourceLine" id="cb14-3" data-line-number="3"></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="co"># content is single specimen object</span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5">res<span class="op">$</span>content</a></code></pre></div>
<pre><code>## &lt;Specimen&gt;
## Fields:
##  sourceSystem:   object of class &lt;SourceSystem&gt; 
##  sourceSystemId:  RMNH.MAM.17209.B 
##  recordURI:   
##  id:  RMNH.MAM.17209.B@CRS 
##  unitID:  RMNH.MAM.17209.B 
##  unitGUID:    http://data.biodiversitydata.nl/naturalis/specimen/RMNH.MAM.17209.B 
##  collectorsFieldNumber:   
##  assemblageID:    
##  sourceInstitutionID:     Naturalis Biodiversity Center 
##  sourceID:    CRS 
##  previousSourceID:    
##  owner:   Naturalis Biodiversity Center 
##  licenseType:     Copyright 
##  license:     CC0 4.0 
##  recordBasis:     PreservedSpecimen 
##  kindOfUnit:  skin 
##  collectionType:  Mammalia 
##  sex:     female 
##  phaseOrStage:    juvenile 
##  title:   RMNH.MAM.17209.b_RMNH_MAM_17209.b 
##  notes:   
##  preparationType:     
##  previousUnitsText:   
##  numberOfSpecimen:    1 
##  fromCaptivity:   FALSE 
##  objectPublic:    TRUE 
##  multiMediaPublic:    TRUE 
##  acquiredFrom:   object of class &lt;Agent&gt; 
##  gatheringEvent: object of class &lt;GatheringEvent&gt; 
##  identifications:    list of length 1 
##  associatedMultiMediaUris:   list of length 0 
##  theme:   
## Methods:
##  fromJSONString
##  toJSONString
##  fromList
##  toList
##  print</code></pre>
</div>
<div id="find_by_ids" class="section level4">
<h4 class="hasAnchor">
<a href="#find_by_ids" class="anchor"></a>find_by_ids</h4>
<p>Same as <code>find</code>, but takes multiple IDs:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">ids &lt;-<span class="st"> "RMNH.INS.657083@CRS,L.1589244@BRAHMS"</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">find_by_ids</span>(ids)</a></code></pre></div>
</div>
<div id="find_by_unit_id" class="section level4">
<h4 class="hasAnchor">
<a href="#find_by_unit_id" class="anchor"></a>find_by_unit_id</h4>
<p>Find a specimen by its <code>unitID</code>:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">unitID &lt;-<span class="st"> "RMNH.MAM.1513"</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2">res &lt;-<span class="st"> </span>sc<span class="op">$</span><span class="kw">find_by_unit_id</span>(unitID)</a></code></pre></div>
</div>
</div>
<div id="aggregation-services" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services" class="anchor"></a>Aggregation services</h3>
<p>Aggregation services group available data according to different criteria.</p>
<div id="get_distinct_values" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values" class="anchor"></a>get_distinct_values</h4>
<p>This method takes a specific field as an argument and returns all possible values and the frequency for that field in the data. Below we get all possible values for the country in which a specimen was collected.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"gatheringEvent.country"</span>)</a></code></pre></div>
<p><strong>Note</strong>: By default, <code>get_distinct_values</code> lists only the first 10 hits. The above query thus does not reflect the distinct values in the hole dataset. This number can be increased with e.g.Â setting the <code>size</code> parameter in a <code>QuerySpec</code> object passed to the method.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"gatheringEvent.country"</span>,</a>
<a class="sourceLine" id="cb19-2" data-line-number="2">                       <span class="dt">querySpec =</span> QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">size =</span> <span class="dv">10000</span>))</a></code></pre></div>
</div>
<div id="count_distinct_values" class="section level4">
<h4 class="hasAnchor">
<a href="#count_distinct_values" class="anchor"></a>count_distinct_values</h4>
<p>Instead of returning all different values for a given field, this method does a mere count:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">sc<span class="op">$</span><span class="kw">count_distinct_values</span>(<span class="st">"gatheringEvent.country"</span>)</a></code></pre></div>
</div>
<div id="get_distinct_values_per_group" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values_per_group" class="anchor"></a>get_distinct_values_per_group</h4>
<p>Suppose you want to add another filter to the above query of retrieving all distinct values for a given field, such as splitting the results by their respective source system:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_distinct_values_per_group</span>(<span class="st">"sourceSystem.name"</span>, <span class="st">"gatheringEvent.country"</span>)</a></code></pre></div>
</div>
<div id="count_distinct_values_per_group" class="section level4">
<h4 class="hasAnchor">
<a href="#count_distinct_values_per_group" class="anchor"></a>count_distinct_values_per_group</h4>
<p>If no return of the actual values is needed, a simple count per group is done as follows:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">sc<span class="op">$</span><span class="kw">count_distinct_values_per_group</span>(<span class="st">"sourceSystem.name"</span>,</a>
<a class="sourceLine" id="cb22-2" data-line-number="2">                                   <span class="st">"gatheringEvent.country"</span>)<span class="op">$</span>content</a></code></pre></div>
</div>
</div>
<div id="metadata-services" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services" class="anchor"></a>Metadata services</h3>
<p>Specimen Metadata services include the same standard metadata services as for the other data types:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="co"># get all paths for the Specimen datatype </span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##   [1] "sourceSystem.code"                                                              
##   [2] "sourceSystem.name"                                                              
##   [3] "sourceSystemId"                                                                 
##   [4] "recordURI"                                                                      
##   [5] "unitID"                                                                         
##   [6] "unitGUID"                                                                       
##   [7] "collectorsFieldNumber"                                                          
##   [8] "assemblageID"                                                                   
##   [9] "sourceInstitutionID"                                                            
##  [10] "sourceID"                                                                       
##  [11] "previousSourceID"                                                               
##  [12] "owner"                                                                          
##  [13] "licenseType"                                                                    
##  [14] "license"                                                                        
##  [15] "recordBasis"                                                                    
##  [16] "kindOfUnit"                                                                     
##  [17] "collectionType"                                                                 
##  [18] "sex"                                                                            
##  [19] "phaseOrStage"                                                                   
##  [20] "title"                                                                          
##  [21] "notes"                                                                          
##  [22] "preparationType"                                                                
##  [23] "previousUnitsText"                                                              
##  [24] "numberOfSpecimen"                                                               
##  [25] "fromCaptivity"                                                                  
##  [26] "objectPublic"                                                                   
##  [27] "multiMediaPublic"                                                               
##  [28] "acquiredFrom.agentText"                                                         
##  [29] "gatheringEvent.projectTitle"                                                    
##  [30] "gatheringEvent.worldRegion"                                                     
##  [31] "gatheringEvent.continent"                                                       
##  [32] "gatheringEvent.country"                                                         
##  [33] "gatheringEvent.iso3166Code"                                                     
##  [34] "gatheringEvent.provinceState"                                                   
##  [35] "gatheringEvent.island"                                                          
##  [36] "gatheringEvent.locality"                                                        
##  [37] "gatheringEvent.city"                                                            
##  [38] "gatheringEvent.sublocality"                                                     
##  [39] "gatheringEvent.localityText"                                                    
##  [40] "gatheringEvent.dateTimeBegin"                                                   
##  [41] "gatheringEvent.dateTimeEnd"                                                     
##  [42] "gatheringEvent.dateText"                                                        
##  [43] "gatheringEvent.method"                                                          
##  [44] "gatheringEvent.altitude"                                                        
##  [45] "gatheringEvent.altitudeUnifOfMeasurement"                                       
##  [46] "gatheringEvent.behavior"                                                        
##  [47] "gatheringEvent.biotopeText"                                                     
##  [48] "gatheringEvent.depth"                                                           
##  [49] "gatheringEvent.depthUnitOfMeasurement"                                          
##  [50] "gatheringEvent.code"                                                            
##  [51] "gatheringEvent.establishmentMeans"                                              
##  [52] "gatheringEvent.gatheringPersons.agentText"                                      
##  [53] "gatheringEvent.gatheringPersons.fullName"                                       
##  [54] "gatheringEvent.gatheringPersons.organization.agentText"                         
##  [55] "gatheringEvent.gatheringPersons.organization.name"                              
##  [56] "gatheringEvent.gatheringOrganizations.agentText"                                
##  [57] "gatheringEvent.gatheringOrganizations.name"                                     
##  [58] "gatheringEvent.siteCoordinates.longitudeDecimal"                                
##  [59] "gatheringEvent.siteCoordinates.latitudeDecimal"                                 
##  [60] "gatheringEvent.siteCoordinates.gridCellSystem"                                  
##  [61] "gatheringEvent.siteCoordinates.gridLatitudeDecimal"                             
##  [62] "gatheringEvent.siteCoordinates.gridLongitudeDecimal"                            
##  [63] "gatheringEvent.siteCoordinates.gridCellCode"                                    
##  [64] "gatheringEvent.siteCoordinates.gridQualifier"                                   
##  [65] "gatheringEvent.siteCoordinates.coordinateErrorDistanceInMeters"                 
##  [66] "gatheringEvent.siteCoordinates.spatialDatum"                                    
##  [67] "gatheringEvent.siteCoordinates.geoShape"                                        
##  [68] "gatheringEvent.namedAreas.areaName"                                             
##  [69] "gatheringEvent.namedAreas.areaClass"                                            
##  [70] "gatheringEvent.associatedTaxa.name"                                             
##  [71] "gatheringEvent.associatedTaxa.relationType"                                     
##  [72] "gatheringEvent.chronoStratigraphy.youngRegionalSubstage"                        
##  [73] "gatheringEvent.chronoStratigraphy.youngRegionalStage"                           
##  [74] "gatheringEvent.chronoStratigraphy.youngRegionalSeries"                          
##  [75] "gatheringEvent.chronoStratigraphy.youngDatingQualifier"                         
##  [76] "gatheringEvent.chronoStratigraphy.youngInternSystem"                            
##  [77] "gatheringEvent.chronoStratigraphy.youngInternSubstage"                          
##  [78] "gatheringEvent.chronoStratigraphy.youngInternStage"                             
##  [79] "gatheringEvent.chronoStratigraphy.youngInternSeries"                            
##  [80] "gatheringEvent.chronoStratigraphy.youngInternErathem"                           
##  [81] "gatheringEvent.chronoStratigraphy.youngInternEonothem"                          
##  [82] "gatheringEvent.chronoStratigraphy.youngChronoName"                              
##  [83] "gatheringEvent.chronoStratigraphy.youngCertainty"                               
##  [84] "gatheringEvent.chronoStratigraphy.oldDatingQualifier"                           
##  [85] "gatheringEvent.chronoStratigraphy.chronoPreferredFlag"                          
##  [86] "gatheringEvent.chronoStratigraphy.oldRegionalSubstage"                          
##  [87] "gatheringEvent.chronoStratigraphy.oldRegionalStage"                             
##  [88] "gatheringEvent.chronoStratigraphy.oldRegionalSeries"                            
##  [89] "gatheringEvent.chronoStratigraphy.oldInternSystem"                              
##  [90] "gatheringEvent.chronoStratigraphy.oldInternSubstage"                            
##  [91] "gatheringEvent.chronoStratigraphy.oldInternStage"                               
##  [92] "gatheringEvent.chronoStratigraphy.oldInternSeries"                              
##  [93] "gatheringEvent.chronoStratigraphy.oldInternErathem"                             
##  [94] "gatheringEvent.chronoStratigraphy.oldInternEonothem"                            
##  [95] "gatheringEvent.chronoStratigraphy.oldChronoName"                                
##  [96] "gatheringEvent.chronoStratigraphy.chronoIdentifier"                             
##  [97] "gatheringEvent.chronoStratigraphy.oldCertainty"                                 
##  [98] "gatheringEvent.bioStratigraphy.youngBioDatingQualifier"                         
##  [99] "gatheringEvent.bioStratigraphy.youngBioName"                                    
## [100] "gatheringEvent.bioStratigraphy.youngFossilZone"                                 
## [101] "gatheringEvent.bioStratigraphy.youngFossilSubZone"                              
## [102] "gatheringEvent.bioStratigraphy.youngBioCertainty"                               
## [103] "gatheringEvent.bioStratigraphy.youngStratType"                                  
## [104] "gatheringEvent.bioStratigraphy.bioDatingQualifier"                              
## [105] "gatheringEvent.bioStratigraphy.bioPreferredFlag"                                
## [106] "gatheringEvent.bioStratigraphy.rangePosition"                                   
## [107] "gatheringEvent.bioStratigraphy.oldBioName"                                      
## [108] "gatheringEvent.bioStratigraphy.bioIdentifier"                                   
## [109] "gatheringEvent.bioStratigraphy.oldFossilzone"                                   
## [110] "gatheringEvent.bioStratigraphy.oldFossilSubzone"                                
## [111] "gatheringEvent.bioStratigraphy.oldBioCertainty"                                 
## [112] "gatheringEvent.bioStratigraphy.oldBioStratType"                                 
## [113] "gatheringEvent.lithoStratigraphy.qualifier"                                     
## [114] "gatheringEvent.lithoStratigraphy.preferredFlag"                                 
## [115] "gatheringEvent.lithoStratigraphy.member2"                                       
## [116] "gatheringEvent.lithoStratigraphy.member"                                        
## [117] "gatheringEvent.lithoStratigraphy.informalName2"                                 
## [118] "gatheringEvent.lithoStratigraphy.informalName"                                  
## [119] "gatheringEvent.lithoStratigraphy.importedName2"                                 
## [120] "gatheringEvent.lithoStratigraphy.importedName1"                                 
## [121] "gatheringEvent.lithoStratigraphy.lithoIdentifier"                               
## [122] "gatheringEvent.lithoStratigraphy.formation2"                                    
## [123] "gatheringEvent.lithoStratigraphy.formationGroup2"                               
## [124] "gatheringEvent.lithoStratigraphy.formationGroup"                                
## [125] "gatheringEvent.lithoStratigraphy.formation"                                     
## [126] "gatheringEvent.lithoStratigraphy.certainty2"                                    
## [127] "gatheringEvent.lithoStratigraphy.certainty"                                     
## [128] "gatheringEvent.lithoStratigraphy.bed2"                                          
## [129] "gatheringEvent.lithoStratigraphy.bed"                                           
## [130] "informationWithheld"                                                            
## [131] "dataGeneralizations"                                                            
## [132] "modified"                                                                       
## [133] "identifications.taxonRank"                                                      
## [134] "identifications.scientificName.fullScientificName"                              
## [135] "identifications.scientificName.taxonomicStatus"                                 
## [136] "identifications.scientificName.genusOrMonomial"                                 
## [137] "identifications.scientificName.subgenus"                                        
## [138] "identifications.scientificName.specificEpithet"                                 
## [139] "identifications.scientificName.infraspecificEpithet"                            
## [140] "identifications.scientificName.infraspecificMarker"                             
## [141] "identifications.scientificName.nameAddendum"                                    
## [142] "identifications.scientificName.authorshipVerbatim"                              
## [143] "identifications.scientificName.author"                                          
## [144] "identifications.scientificName.year"                                            
## [145] "identifications.scientificName.scientificNameGroup"                             
## [146] "identifications.scientificName.references.titleCitation"                        
## [147] "identifications.scientificName.references.citationDetail"                       
## [148] "identifications.scientificName.references.uri"                                  
## [149] "identifications.scientificName.references.author.agentText"                     
## [150] "identifications.scientificName.references.author.fullName"                      
## [151] "identifications.scientificName.references.author.organization.agentText"        
## [152] "identifications.scientificName.references.author.organization.name"             
## [153] "identifications.scientificName.references.publicationDate"                      
## [154] "identifications.scientificName.experts.agentText"                               
## [155] "identifications.scientificName.experts.fullName"                                
## [156] "identifications.scientificName.experts.organization.agentText"                  
## [157] "identifications.scientificName.experts.organization.name"                       
## [158] "identifications.typeStatus"                                                     
## [159] "identifications.dateIdentified"                                                 
## [160] "identifications.defaultClassification.kingdom"                                  
## [161] "identifications.defaultClassification.phylum"                                   
## [162] "identifications.defaultClassification.className"                                
## [163] "identifications.defaultClassification.order"                                    
## [164] "identifications.defaultClassification.superFamily"                              
## [165] "identifications.defaultClassification.family"                                   
## [166] "identifications.defaultClassification.genus"                                    
## [167] "identifications.defaultClassification.subgenus"                                 
## [168] "identifications.defaultClassification.specificEpithet"                          
## [169] "identifications.defaultClassification.infraspecificEpithet"                     
## [170] "identifications.defaultClassification.infraspecificRank"                        
## [171] "identifications.systemClassification.rank"                                      
## [172] "identifications.systemClassification.name"                                      
## [173] "identifications.vernacularNames.name"                                           
## [174] "identifications.vernacularNames.language"                                       
## [175] "identifications.vernacularNames.preferred"                                      
## [176] "identifications.vernacularNames.references.titleCitation"                       
## [177] "identifications.vernacularNames.references.citationDetail"                      
## [178] "identifications.vernacularNames.references.uri"                                 
## [179] "identifications.vernacularNames.references.author.agentText"                    
## [180] "identifications.vernacularNames.references.author.fullName"                     
## [181] "identifications.vernacularNames.references.author.organization.agentText"       
## [182] "identifications.vernacularNames.references.author.organization.name"            
## [183] "identifications.vernacularNames.references.publicationDate"                     
## [184] "identifications.vernacularNames.experts.agentText"                              
## [185] "identifications.vernacularNames.experts.fullName"                               
## [186] "identifications.vernacularNames.experts.organization.agentText"                 
## [187] "identifications.vernacularNames.experts.organization.name"                      
## [188] "identifications.identificationQualifiers"                                       
## [189] "identifications.identifiers.agentText"                                          
## [190] "identifications.taxonomicEnrichments.vernacularNames.name"                      
## [191] "identifications.taxonomicEnrichments.vernacularNames.language"                  
## [192] "identifications.taxonomicEnrichments.synonyms.fullScientificName"               
## [193] "identifications.taxonomicEnrichments.synonyms.taxonomicStatus"                  
## [194] "identifications.taxonomicEnrichments.synonyms.genusOrMonomial"                  
## [195] "identifications.taxonomicEnrichments.synonyms.subgenus"                         
## [196] "identifications.taxonomicEnrichments.synonyms.specificEpithet"                  
## [197] "identifications.taxonomicEnrichments.synonyms.infraspecificEpithet"             
## [198] "identifications.taxonomicEnrichments.synonyms.authorshipVerbatim"               
## [199] "identifications.taxonomicEnrichments.defaultClassification.kingdom"             
## [200] "identifications.taxonomicEnrichments.defaultClassification.phylum"              
## [201] "identifications.taxonomicEnrichments.defaultClassification.className"           
## [202] "identifications.taxonomicEnrichments.defaultClassification.order"               
## [203] "identifications.taxonomicEnrichments.defaultClassification.superFamily"         
## [204] "identifications.taxonomicEnrichments.defaultClassification.family"              
## [205] "identifications.taxonomicEnrichments.defaultClassification.genus"               
## [206] "identifications.taxonomicEnrichments.defaultClassification.subgenus"            
## [207] "identifications.taxonomicEnrichments.defaultClassification.specificEpithet"     
## [208] "identifications.taxonomicEnrichments.defaultClassification.infraspecificEpithet"
## [209] "identifications.taxonomicEnrichments.defaultClassification.infraspecificRank"   
## [210] "identifications.taxonomicEnrichments.sourceSystem.code"                         
## [211] "identifications.taxonomicEnrichments.taxonId"                                   
## [212] "identifications.preferred"                                                      
## [213] "identifications.verificationStatus"                                             
## [214] "identifications.rockType"                                                       
## [215] "identifications.associatedFossilAssemblage"                                     
## [216] "identifications.rockMineralUsage"                                               
## [217] "identifications.associatedMineralName"                                          
## [218] "identifications.remarks"                                                        
## [219] "associatedMultiMediaUris.accessUri"                                             
## [220] "associatedMultiMediaUris.format"                                                
## [221] "associatedMultiMediaUris.variant"                                               
## [222] "theme"</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="co"># get info e.g. for field collectionType</span></a>
<a class="sourceLine" id="cb25-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content<span class="op">$</span>collectionType</a></code></pre></div>
<pre><code>## $indexed
## [1] TRUE
## 
## $type
## [1] "keyword"
## 
## $allowedOperators
##  [1] "EQUALS"             "NOT_EQUALS"         "EQUALS_IC"         
##  [4] "NOT_EQUALS_IC"      "IN"                 "NOT_IN"            
##  [7] "MATCHES"            "NOT_MATCHES"        "STARTS_WITH"       
## [10] "NOT_STARTS_WITH"    "STARTS_WITH_IC"     "NOT_STARTS_WITH_IC"</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="co"># get all settings</span></a>
<a class="sourceLine" id="cb27-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_settings</span>()</a></code></pre></div>
<pre><code>## &lt;Response&gt;
## Fields:
##  content: object of class &lt;list&gt; 
##  response: object of class &lt;response&gt; (httr package)</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="co"># get specific setting</span></a>
<a class="sourceLine" id="cb29-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="co"># check if operator is allowed</span></a>
<a class="sourceLine" id="cb31-2" data-line-number="2">sc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"gatheringEvent.continent"</span>, <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div id="available-paths-and-fields" class="section level4">
<h4 class="hasAnchor">
<a href="#available-paths-and-fields" class="anchor"></a>Available paths and fields</h4>
<p>All fields can be retrieved with <code>get_paths</code> and specific information on the fields, such as allowed operators etc. with <code>get_field_info</code></p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a>
<a class="sourceLine" id="cb33-2" data-line-number="2">sc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content</a></code></pre></div>
</div>
<div id="settings" class="section level4">
<h4 class="hasAnchor">
<a href="#settings" class="anchor"></a>Settings</h4>
<p>Multimedia-specific settings can be retrieved with <code>get_settings</code> and a specific setting with <code>get_setting</code>:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_settings</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $index.max_result_window
## [1] 50000
## 
## $specimen.group_by_scientific_name.max_num_buckets
## [1] 10000</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1">sc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
</div>
<div id="operators" class="section level4">
<h4 class="hasAnchor">
<a href="#operators" class="anchor"></a>Operators</h4>
<p>To test if a certain operator can be used for a multimedia query:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1">sc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"identifications.defaultClassification.genus"</span>,</a>
<a class="sourceLine" id="cb38-2" data-line-number="2">                       <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="dwca-download-services" class="section level3">
<h3 class="hasAnchor">
<a href="#dwca-download-services" class="anchor"></a>DwCA download services</h3>
<p>In addition to query services that return JSON formatted data, the NBA also offers the export of <a href="https://en.wikipedia.org/wiki/Darwin_Core_Archive">Darwin Core Archive (DwCA)</a> files. These files are by default zip archives, please refer to our <a href="https://en.wikipedia.org/wiki/Darwin_Core_Archive">official API documentation</a> for more information.</p>
<div id="static-download" class="section level4">
<h4 class="hasAnchor">
<a href="#static-download" class="anchor"></a>Static download</h4>
<p>Static download services offer the download of predefined datasets. The sets that are available for download can be queried with <code>dwca_get_data_set_names</code>:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1">sc<span class="op">$</span><span class="kw">dwca_get_data_set_names</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "amphibia-and-reptilia"      "aves"                      
##  [3] "birdsounds"                 "botany"                    
##  [5] "brachiopoda"                "cainozoic-mollusca"        
##  [7] "chelicerata-and-myriapoda"  "cnidaria"                  
##  [9] "coleoptera"                 "collembola"                
## [11] "crustacea"                  "diptera"                   
## [13] "echinodermata"              "foraminifera"              
## [15] "hemiptera"                  "hymenoptera"               
## [17] "lepidoptera"                "mammalia"                  
## [19] "micropaleontology"          "mollusca"                  
## [21] "observations"               "odonata"                   
## [23] "orthopteroidea"             "paleobotany"               
## [25] "paleontology-invertebrates" "pisces"                    
## [27] "porifera"                   "tunicata"</code></pre>
<p>A dataset can then be downloaded using <code>dwca_get_data_set</code>. A filename can be given as argument, if none is given, the DwCA archive is written to <em>download-YYYY-MM-DDThh:mm.zip</em> in the current working directory.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="co"># download dataset 'porifera' to temporary file</span></a>
<a class="sourceLine" id="cb42-2" data-line-number="2">filename &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempfile</a></span>(<span class="dt">fileext=</span><span class="st">".zip"</span>)</a>
<a class="sourceLine" id="cb42-3" data-line-number="3">sc<span class="op">$</span><span class="kw">dwca_get_data_set</span>(<span class="st">'porifera'</span>, <span class="dt">filename=</span>filename)                                                                               </a></code></pre></div>
<pre><code>## Query result written to /tmp/RtmpSAW22y/file7a832fd2eb64.zip</code></pre>
</div>
<div id="dynamic-download" class="section level4">
<h4 class="hasAnchor">
<a href="#dynamic-download" class="anchor"></a>Dynamic download</h4>
<p>The dynamic download function <code>dwca_query</code> allows for download of arbitrary sets, defined by the userâ€™s query. The arguments to this methods are similar to <code>query</code>, plus the filename:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="co"># download all specimen of genus 'Hydrochoerus'</span></a>
<a class="sourceLine" id="cb44-2" data-line-number="2">filename &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempfile</a></span>(<span class="dt">fileext =</span> <span class="st">".zip"</span>)</a>
<a class="sourceLine" id="cb44-3" data-line-number="3">qs &lt;-</a>
<a class="sourceLine" id="cb44-4" data-line-number="4"><span class="st">  </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</a>
<a class="sourceLine" id="cb44-5" data-line-number="5">    QueryCondition<span class="op">$</span><span class="kw">new</span>(</a>
<a class="sourceLine" id="cb44-6" data-line-number="6">      <span class="dt">field =</span> <span class="st">"identifications.defaultClassification.genus"</span>,</a>
<a class="sourceLine" id="cb44-7" data-line-number="7">      <span class="dt">operator =</span> <span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb44-8" data-line-number="8">      <span class="dt">value =</span> <span class="st">"Hydrochoerus"</span></a>
<a class="sourceLine" id="cb44-9" data-line-number="9">    )</a>
<a class="sourceLine" id="cb44-10" data-line-number="10">  ))</a>
<a class="sourceLine" id="cb44-11" data-line-number="11">sc<span class="op">$</span><span class="kw">dwca_query</span>(<span class="dt">querySpec =</span> qs, <span class="dt">filename =</span> filename)</a></code></pre></div>
<pre><code>## Query result written to /tmp/RtmpSAW22y/file7a83251e5518.zip</code></pre>
</div>
</div>
</div>
<div id="taxonomic-data-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#taxonomic-data-services" class="anchor"></a>Taxonomic data services</h2>
<div id="query-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services-1" class="anchor"></a>Query Services</h3>
<div id="query-1" class="section level4">
<h4 class="hasAnchor">
<a href="#query-1" class="anchor"></a>query</h4>
<p>Query for taxon document with given search criteria. Example:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1"><span class="co"># query for taxa of genus 'Sedum' that are in the Netherlands Soortenregister (NSR)</span></a>
<a class="sourceLine" id="cb46-2" data-line-number="2">tc &lt;-<span class="st"> </span>TaxonClient<span class="op">$</span><span class="kw">new</span>()</a>
<a class="sourceLine" id="cb46-3" data-line-number="3">qc &lt;-</a>
<a class="sourceLine" id="cb46-4" data-line-number="4"><span class="st">  </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span> <span class="st">"acceptedName.genusOrMonomial"</span>,</a>
<a class="sourceLine" id="cb46-5" data-line-number="5">                     <span class="dt">operator =</span> <span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb46-6" data-line-number="6">                     <span class="dt">value =</span> <span class="st">"Sedum"</span>)</a>
<a class="sourceLine" id="cb46-7" data-line-number="7">qc2 &lt;-</a>
<a class="sourceLine" id="cb46-8" data-line-number="8"><span class="st">  </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span> <span class="st">"sourceSystem.code"</span>,</a>
<a class="sourceLine" id="cb46-9" data-line-number="9">                     <span class="dt">operator =</span> <span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb46-10" data-line-number="10">                     <span class="dt">value =</span> <span class="st">"NSR"</span>)</a>
<a class="sourceLine" id="cb46-11" data-line-number="11">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(qc, qc2))</a>
<a class="sourceLine" id="cb46-12" data-line-number="12">tc<span class="op">$</span><span class="kw">query</span>(qs)</a></code></pre></div>
</div>
<div id="download_query-1" class="section level4">
<h4 class="hasAnchor">
<a href="#download_query-1" class="anchor"></a>download_query</h4>
<p>The <code>query</code> function is limited to retrieve 50000 taxa at once (this is determined in the parameter <code>index.max_result_window</code>, the value is retrievable using the <code>getSettings</code> method in the metadata section). In order to provide access for a larger amount of data, the <code>query_download</code> takes the same arguments as <code>query</code>, but download the data as a gzip stream under the hood. Unlike <code>query</code>, <code>query_download</code> returns a list of taxon objects instead of a <code>ResultSet</code>.</p>
</div>
</div>
<div id="data-access-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services-1" class="anchor"></a>Data access services</h3>
<div id="count-1" class="section level4">
<h4 class="hasAnchor">
<a href="#count-1" class="anchor"></a>count</h4>
<p>For a given query, do not return <code>Taxon</code> objects but the mere count. Example</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1"><span class="co"># get counts for taxa of genus 'Sedum' that are in the</span></a>
<a class="sourceLine" id="cb47-2" data-line-number="2"><span class="co">#  Netherlands Soortenregister (NSR)</span></a>
<a class="sourceLine" id="cb47-3" data-line-number="3">qc &lt;-</a>
<a class="sourceLine" id="cb47-4" data-line-number="4"><span class="st">  </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span> <span class="st">"acceptedName.genusOrMonomial"</span>,</a>
<a class="sourceLine" id="cb47-5" data-line-number="5">                     <span class="dt">operator =</span> <span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb47-6" data-line-number="6">                     <span class="dt">value =</span> <span class="st">"Sedum"</span>)</a>
<a class="sourceLine" id="cb47-7" data-line-number="7">qc2 &lt;-</a>
<a class="sourceLine" id="cb47-8" data-line-number="8"><span class="st">  </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span> <span class="st">"sourceSystem.code"</span>,</a>
<a class="sourceLine" id="cb47-9" data-line-number="9">                     <span class="dt">operator =</span> <span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb47-10" data-line-number="10">                     <span class="dt">value =</span> <span class="st">"NSR"</span>)</a>
<a class="sourceLine" id="cb47-11" data-line-number="11">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(qc, qc2))</a>
<a class="sourceLine" id="cb47-12" data-line-number="12">tc<span class="op">$</span><span class="kw">count</span>(qs)</a></code></pre></div>
<pre><code>## &lt;Response&gt;
## Fields:
##  content: object of class &lt;integer&gt; 
##  response: object of class &lt;response&gt; (httr package)</code></pre>
</div>
<div id="find-1" class="section level4">
<h4 class="hasAnchor">
<a href="#find-1" class="anchor"></a>find</h4>
<p>Returns a taxon object given its identifier:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1">tc<span class="op">$</span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/apropos">find</a></span>(<span class="st">"27706109@COL"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## No encoding supplied: defaulting to UTF-8.
## No encoding supplied: defaulting to UTF-8.
## No encoding supplied: defaulting to UTF-8.</code></pre>
<pre><code>## Warning in self$handleError(response): 404 (NOT FOUND)
## No Specimen exists with ID 27706109@COL</code></pre>
<pre><code>## [1] "404 (NOT FOUND)\nNo Specimen exists with ID 27706109@COL"</code></pre>
</div>
<div id="find_by_ids-1" class="section level4">
<h4 class="hasAnchor">
<a href="#find_by_ids-1" class="anchor"></a>find_by_ids</h4>
<p>Given a string with comma-separated identifiers, returns a list of taxon objects:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" data-line-number="1">ids &lt;-<span class="st"> "27706109@COL,27704140@COL,27706110@COL,27706111@COL,27706108@COL"</span></a>
<a class="sourceLine" id="cb53-2" data-line-number="2">res &lt;-<span class="st"> </span>tc<span class="op">$</span><span class="kw">find_by_ids</span>(ids)</a></code></pre></div>
</div>
</div>
<div id="aggregation-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services-1" class="anchor"></a>Aggregation services</h3>
<div id="get_distinct_values-1" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values-1" class="anchor"></a>get_distinct_values</h4>
<p>This method takes a specific field as an argument and returns all possible values and the frequency for that field in the data. Example: get all data source systems for taxon objects:</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1">tc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"sourceSystem.name"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $`Species 2000 - Catalogue Of Life`
## [1] 1962192
## 
## $`Naturalis - Dutch Species Register`
## [1] 47646</code></pre>
</div>
</div>
<div id="metadata-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services-1" class="anchor"></a>Metadata services</h3>
<p>Taxon Metadata services include the same standard metadata services as for the other data types:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" data-line-number="1"><span class="co"># get all paths for the Taxon datatype </span></a>
<a class="sourceLine" id="cb56-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "sourceSystem.code"                                       
##  [2] "sourceSystem.name"                                       
##  [3] "sourceSystemId"                                          
##  [4] "recordURI"                                               
##  [5] "sourceSystemParentId"                                    
##  [6] "taxonRank"                                               
##  [7] "taxonRemarks"                                            
##  [8] "occurrenceStatusVerbatim"                                
##  [9] "acceptedName.fullScientificName"                         
## [10] "acceptedName.taxonomicStatus"                            
## [11] "acceptedName.genusOrMonomial"                            
## [12] "acceptedName.subgenus"                                   
## [13] "acceptedName.specificEpithet"                            
## [14] "acceptedName.infraspecificEpithet"                       
## [15] "acceptedName.infraspecificMarker"                        
## [16] "acceptedName.nameAddendum"                               
## [17] "acceptedName.authorshipVerbatim"                         
## [18] "acceptedName.author"                                     
## [19] "acceptedName.year"                                       
## [20] "acceptedName.scientificNameGroup"                        
## [21] "acceptedName.references.titleCitation"                   
## [22] "acceptedName.references.citationDetail"                  
## [23] "acceptedName.references.uri"                             
## [24] "acceptedName.references.author.agentText"                
## [25] "acceptedName.references.author.fullName"                 
## [26] "acceptedName.references.author.organization.agentText"   
## [27] "acceptedName.references.author.organization.name"        
## [28] "acceptedName.references.publicationDate"                 
## [29] "acceptedName.experts.agentText"                          
## [30] "acceptedName.experts.fullName"                           
## [31] "acceptedName.experts.organization.agentText"             
## [32] "acceptedName.experts.organization.name"                  
## [33] "defaultClassification.kingdom"                           
## [34] "defaultClassification.phylum"                            
## [35] "defaultClassification.className"                         
## [36] "defaultClassification.order"                             
## [37] "defaultClassification.superFamily"                       
## [38] "defaultClassification.family"                            
## [39] "defaultClassification.genus"                             
## [40] "defaultClassification.subgenus"                          
## [41] "defaultClassification.specificEpithet"                   
## [42] "defaultClassification.infraspecificEpithet"              
## [43] "defaultClassification.infraspecificRank"                 
## [44] "systemClassification.rank"                               
## [45] "systemClassification.name"                               
## [46] "synonyms.fullScientificName"                             
## [47] "synonyms.taxonomicStatus"                                
## [48] "synonyms.genusOrMonomial"                                
## [49] "synonyms.subgenus"                                       
## [50] "synonyms.specificEpithet"                                
## [51] "synonyms.infraspecificEpithet"                           
## [52] "synonyms.infraspecificMarker"                            
## [53] "synonyms.nameAddendum"                                   
## [54] "synonyms.authorshipVerbatim"                             
## [55] "synonyms.author"                                         
## [56] "synonyms.year"                                           
## [57] "synonyms.scientificNameGroup"                            
## [58] "synonyms.references.titleCitation"                       
## [59] "synonyms.references.citationDetail"                      
## [60] "synonyms.references.uri"                                 
## [61] "synonyms.references.author.agentText"                    
## [62] "synonyms.references.author.fullName"                     
## [63] "synonyms.references.author.organization.agentText"       
## [64] "synonyms.references.author.organization.name"            
## [65] "synonyms.references.publicationDate"                     
## [66] "synonyms.experts.agentText"                              
## [67] "synonyms.experts.fullName"                               
## [68] "synonyms.experts.organization.agentText"                 
## [69] "synonyms.experts.organization.name"                      
## [70] "vernacularNames.name"                                    
## [71] "vernacularNames.language"                                
## [72] "vernacularNames.preferred"                               
## [73] "vernacularNames.references.titleCitation"                
## [74] "vernacularNames.references.citationDetail"               
## [75] "vernacularNames.references.uri"                          
## [76] "vernacularNames.references.author.agentText"             
## [77] "vernacularNames.references.author.fullName"              
## [78] "vernacularNames.references.author.organization.agentText"
## [79] "vernacularNames.references.author.organization.name"     
## [80] "vernacularNames.references.publicationDate"              
## [81] "vernacularNames.experts.agentText"                       
## [82] "vernacularNames.experts.fullName"                        
## [83] "vernacularNames.experts.organization.agentText"          
## [84] "vernacularNames.experts.organization.name"               
## [85] "descriptions.category"                                   
## [86] "descriptions.description"                                
## [87] "descriptions.language"                                   
## [88] "references.titleCitation"                                
## [89] "references.citationDetail"                               
## [90] "references.uri"                                          
## [91] "references.author.agentText"                             
## [92] "references.author.fullName"                              
## [93] "references.author.organization.agentText"                
## [94] "references.author.organization.name"                     
## [95] "references.publicationDate"                              
## [96] "experts.agentText"                                       
## [97] "experts.fullName"                                        
## [98] "experts.organization.agentText"                          
## [99] "experts.organization.name"</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" data-line-number="1"><span class="co"># get info e.g. for field collectionType</span></a>
<a class="sourceLine" id="cb58-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content<span class="op">$</span>synonyms.author</a></code></pre></div>
<pre><code>## $indexed
## [1] TRUE
## 
## $type
## [1] "keyword"
## 
## $allowedOperators
##  [1] "EQUALS"             "NOT_EQUALS"         "EQUALS_IC"         
##  [4] "NOT_EQUALS_IC"      "CONTAINS"           "NOT_CONTAINS"      
##  [7] "IN"                 "NOT_IN"             "MATCHES"           
## [10] "NOT_MATCHES"        "STARTS_WITH"        "NOT_STARTS_WITH"   
## [13] "STARTS_WITH_IC"     "NOT_STARTS_WITH_IC"</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb60-1" data-line-number="1"><span class="co"># get all settings</span></a>
<a class="sourceLine" id="cb60-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_settings</span>()</a></code></pre></div>
<pre><code>## &lt;Response&gt;
## Fields:
##  content: object of class &lt;list&gt; 
##  response: object of class &lt;response&gt; (httr package)</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb62-1" data-line-number="1"><span class="co"># get specific setting</span></a>
<a class="sourceLine" id="cb62-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb64-1" data-line-number="1"><span class="co"># check if operator is allowed</span></a>
<a class="sourceLine" id="cb64-2" data-line-number="2">tc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"synonyms.author"</span>, <span class="st">"EQUALS"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div id="available-paths-and-fields-1" class="section level4">
<h4 class="hasAnchor">
<a href="#available-paths-and-fields-1" class="anchor"></a>Available paths and fields</h4>
<p>All fields can be retrieved with <code>get_paths</code> and specific information on the fields, such as allowed operators etc. with <code>get_field_info</code></p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb66-1" data-line-number="1">tc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a>
<a class="sourceLine" id="cb66-2" data-line-number="2">tc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content</a></code></pre></div>
</div>
<div id="settings-1" class="section level4">
<h4 class="hasAnchor">
<a href="#settings-1" class="anchor"></a>Settings</h4>
<p>Multimedia-specific settings can be retrieved with <code>get_settings</code> and a specific setting with <code>get_setting</code>:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" data-line-number="1">tc<span class="op">$</span><span class="kw">get_settings</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $index.max_result_window
## [1] 50000
## 
## $taxon.group_by_scientific_name.max_num_buckets
## [1] 10000</code></pre>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb69-1" data-line-number="1">tc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
</div>
<div id="operators-1" class="section level4">
<h4 class="hasAnchor">
<a href="#operators-1" class="anchor"></a>Operators</h4>
<p>To test if a certain operator can be used for a mutimedia query:</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb71-1" data-line-number="1">tc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"identifications.defaultClassification.genus"</span>,</a>
<a class="sourceLine" id="cb71-2" data-line-number="2">                       <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## Warning in self$handleError(response): Status code:500
## Internal Server Error
## Exception: Invalid element "identifications" in path "identifications.defaultClassification.genus"
## Exception type: nl.naturalis.nba.api.NoSuchFieldException
## Full stack trace stored in response object</code></pre>
<pre><code>## [1] "Internal Server Error"</code></pre>
</div>
</div>
<div id="dwca-download-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#dwca-download-services-1" class="anchor"></a>DwCA download services</h3>
<p>The taxonomic information in the NBA is also available as Darwin-Core archive files.</p>
<div id="static-download-1" class="section level4">
<h4 class="hasAnchor">
<a href="#static-download-1" class="anchor"></a>Static download</h4>
<p>Static download services offer the download of predefined datasets. The sets that are available for download can be queried with <code>dwca_get_data_set_names</code>:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb74-1" data-line-number="1">tc<span class="op">$</span><span class="kw">dwca_get_data_set_names</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "nsr"</code></pre>
<p>A dataset can then be downloaded using <code>dwca_get_data_set</code>. A filename can be given as argument, if none is given, the DwCA archive is written to <em>download-YYYY-MM-DDThh:mm.zip</em> in the current working directory.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb76-1" data-line-number="1"><span class="co"># download dataset 'nsr' to temporary file</span></a>
<a class="sourceLine" id="cb76-2" data-line-number="2">filename &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempfile</a></span>(<span class="dt">fileext=</span><span class="st">".zip"</span>)</a>
<a class="sourceLine" id="cb76-3" data-line-number="3">tc<span class="op">$</span><span class="kw">dwca_get_data_set</span>(<span class="st">'nsr'</span>, <span class="dt">filename=</span>filename)                                                                               </a></code></pre></div>
<pre><code>## Query result written to /tmp/RtmpSAW22y/file7a8359734bf5.zip</code></pre>
</div>
<div id="dynamic-download-1" class="section level4">
<h4 class="hasAnchor">
<a href="#dynamic-download-1" class="anchor"></a>Dynamic download</h4>
<p>The dynamic download function <code>dwca_query</code> allows for download of arbitrary sets, defined by the userâ€™s query. The arguments to this methods are similar to <code>query</code>, plus the filename:</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb78-1" data-line-number="1"><span class="co"># download all taxa for genus 'Clematis'</span></a>
<a class="sourceLine" id="cb78-2" data-line-number="2">filename &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempfile</a></span>(<span class="dt">fileext =</span> <span class="st">".zip"</span>)</a>
<a class="sourceLine" id="cb78-3" data-line-number="3">qs &lt;-</a>
<a class="sourceLine" id="cb78-4" data-line-number="4"><span class="st">  </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</a>
<a class="sourceLine" id="cb78-5" data-line-number="5">    QueryCondition<span class="op">$</span><span class="kw">new</span>(</a>
<a class="sourceLine" id="cb78-6" data-line-number="6">      <span class="dt">field =</span> <span class="st">"defaultClassification.genus"</span>,</a>
<a class="sourceLine" id="cb78-7" data-line-number="7">      <span class="dt">operator =</span> <span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb78-8" data-line-number="8">      <span class="dt">value =</span> <span class="st">"Clematis"</span></a>
<a class="sourceLine" id="cb78-9" data-line-number="9">    )</a>
<a class="sourceLine" id="cb78-10" data-line-number="10">  ))</a>
<a class="sourceLine" id="cb78-11" data-line-number="11">tc<span class="op">$</span><span class="kw">dwca_query</span>(<span class="dt">querySpec =</span> qs, <span class="dt">filename =</span> filename)</a></code></pre></div>
<pre><code>## Query result written to /tmp/RtmpSAW22y/file7a833d3e007d.zip</code></pre>
</div>
</div>
</div>
<div id="geographic-data-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#geographic-data-services" class="anchor"></a>Geographic data services</h2>
<div id="query-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services-2" class="anchor"></a>Query Services</h3>
<p>The GeoArea query service allows for detailed search within the fields of a GeoArea object. As for the other data types, query parameters can be either given as a <code>list</code> or as a <code>QuerySpec</code> object. Below, we make a simple query to get the <code>GeoArea</code> object for the Netherlands:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb80-1" data-line-number="1"><span class="co"># instantiate client for geo areas</span></a>
<a class="sourceLine" id="cb80-2" data-line-number="2">gc &lt;-<span class="st"> </span>GeoClient<span class="op">$</span><span class="kw">new</span>()</a>
<a class="sourceLine" id="cb80-3" data-line-number="3"></a>
<a class="sourceLine" id="cb80-4" data-line-number="4"><span class="co"># query for GeoArea of the Netherlands</span></a>
<a class="sourceLine" id="cb80-5" data-line-number="5">qc &lt;-</a>
<a class="sourceLine" id="cb80-6" data-line-number="6"><span class="st">  </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span> <span class="st">"locality"</span>,</a>
<a class="sourceLine" id="cb80-7" data-line-number="7">                     <span class="dt">operator =</span> <span class="st">"EQUALS"</span>,</a>
<a class="sourceLine" id="cb80-8" data-line-number="8">                     <span class="dt">value =</span> <span class="st">"Netherlands"</span>)</a>
<a class="sourceLine" id="cb80-9" data-line-number="9">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(qc))</a>
<a class="sourceLine" id="cb80-10" data-line-number="10">res &lt;-<span class="st"> </span>gc<span class="op">$</span><span class="kw">query</span>(qs)</a>
<a class="sourceLine" id="cb80-11" data-line-number="11"></a>
<a class="sourceLine" id="cb80-12" data-line-number="12"><span class="co"># get item</span></a>
<a class="sourceLine" id="cb80-13" data-line-number="13">res<span class="op">$</span>content<span class="op">$</span>resultSet[[<span class="dv">1</span>]]<span class="op">$</span>item</a></code></pre></div>
<pre><code>## &lt;GeoArea&gt;
## Fields:
##  sourceSystem:   object of class &lt;SourceSystem&gt; 
##  sourceSystemId:  1004050 
##  recordURI:   
##  id:  1004050@GEO 
##  areaType:    Country 
##  locality:    Netherlands 
##  shape:  list of length 2 
##  source:  World Countries 
##  isoCode:     NLD 
##  countryNL:   Nederland 
## Methods:
##  fromJSONString
##  toJSONString
##  fromList
##  toList
##  print</code></pre>
</div>
<div id="data-access-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services-2" class="anchor"></a>Data access services</h3>
<div id="get_geo_json_for_locality" class="section level4">
<h4 class="hasAnchor">
<a href="#get_geo_json_for_locality" class="anchor"></a>get_geo_json_for_locality</h4>
<p>This is a convenience function to directly extract the a GeoJSON object for a specific locality. <a href="https://en.wikipedia.org/wiki/GeoJSON">GeoJSON</a> is a popular format for storing geographical point- and polygon data. To e.g.Â extract the GeoJSON polygon representation for the Netherlands:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb82-1" data-line-number="1">loc &lt;-<span class="st"> "Netherlands"</span></a>
<a class="sourceLine" id="cb82-2" data-line-number="2">res &lt;-<span class="st"> </span>gc<span class="op">$</span><span class="kw">get_geo_json_for_locality</span>(loc)</a></code></pre></div>
<p>Results are returned as a <code>list</code> by default, but can be easily converted to a JSON string, e.g.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb83-1" data-line-number="1">jsonlite<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jsonlite/topics/fromJSON">toJSON</a></span>(res<span class="op">$</span>content)</a></code></pre></div>
</div>
<div id="count-2" class="section level4">
<h4 class="hasAnchor">
<a href="#count-2" class="anchor"></a>count</h4>
<p>For any given query (with <code>QuerySpec</code> or not), returns the count of matches instead of <code>GeoArea</code> objects:</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb84-1" data-line-number="1"><span class="co"># return count of all GeoAreas</span></a>
<a class="sourceLine" id="cb84-2" data-line-number="2">res &lt;-<span class="st"> </span>gc<span class="op">$</span><span class="kw">count</span>()</a>
<a class="sourceLine" id="cb84-3" data-line-number="3">res<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] 1654</code></pre>
</div>
</div>
<div id="aggregation-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services-2" class="anchor"></a>Aggregation services</h3>
<div id="get_distinct_values-2" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values-2" class="anchor"></a>get_distinct_values</h4>
<p>This function returns all values present for a certain field, and their counts:</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb86-1" data-line-number="1">gc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"areaType"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $Nature
## [1] 987
## 
## $Municipality
## [1] 408
## 
## $Country
## [1] 247
## 
## $Province
## [1] 12</code></pre>
</div>
</div>
<div id="metadata-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services-2" class="anchor"></a>Metadata services</h3>
<p>Geo Metadata services include the same standard metadata services as for the other data types:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb88-1" data-line-number="1"><span class="co"># get all paths for the GeoArea datatype </span></a>
<a class="sourceLine" id="cb88-2" data-line-number="2">gc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "sourceSystem.code" "sourceSystem.name" "sourceSystemId"   
##  [4] "recordURI"         "areaType"          "locality"         
##  [7] "shape"             "source"            "isoCode"          
## [10] "countryNL"</code></pre>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb90-1" data-line-number="1"><span class="co"># get info e.g. for field 'areaType'</span></a>
<a class="sourceLine" id="cb90-2" data-line-number="2">gc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content<span class="op">$</span>areaType</a></code></pre></div>
<pre><code>## $indexed
## [1] TRUE
## 
## $type
## [1] "keyword"
## 
## $allowedOperators
##  [1] "EQUALS"             "NOT_EQUALS"         "EQUALS_IC"         
##  [4] "NOT_EQUALS_IC"      "IN"                 "NOT_IN"            
##  [7] "MATCHES"            "NOT_MATCHES"        "STARTS_WITH"       
## [10] "NOT_STARTS_WITH"    "STARTS_WITH_IC"     "NOT_STARTS_WITH_IC"</code></pre>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb92-1" data-line-number="1"><span class="co"># get all settings</span></a>
<a class="sourceLine" id="cb92-2" data-line-number="2">gc<span class="op">$</span><span class="kw">get_settings</span>()</a></code></pre></div>
<pre><code>## &lt;Response&gt;
## Fields:
##  content: object of class &lt;list&gt; 
##  response: object of class &lt;response&gt; (httr package)</code></pre>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb94-1" data-line-number="1"><span class="co"># get specific setting</span></a>
<a class="sourceLine" id="cb94-2" data-line-number="2">gc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb96-1" data-line-number="1"><span class="co"># check if operator is allowed</span></a>
<a class="sourceLine" id="cb96-2" data-line-number="2">gc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"locality"</span>, <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="multimedia-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#multimedia-services" class="anchor"></a>Multimedia services</h2>
<p>Multimedia services are accessible with a <code>MultimediaClient</code>, instantiated as follows:</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb98-1" data-line-number="1">mc &lt;-<span class="st"> </span>MultimediaClient<span class="op">$</span><span class="kw">new</span>()</a></code></pre></div>
<div id="query-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services-3" class="anchor"></a>Query Services</h3>
<p>As for the other data types, the <code>query</code> method enables simple and complex queries using a list or a <code>QuerySpec</code> object to specify query parameters.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb99-1" data-line-number="1"><span class="co"># example of multimedia query passing parameters as a list</span></a>
<a class="sourceLine" id="cb99-2" data-line-number="2">mc<span class="op">$</span><span class="kw">query</span>(<span class="dt">queryParams =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">collectionType =</span> <span class="st">'Cnidaria'</span>))<span class="op">$</span>content</a></code></pre></div>
<pre><code>## &lt;QueryResult&gt;
## Fields:
##  totalSize:   42350 
##  resultSet:  list of length 10 
## Methods:
##  fromJSONString
##  toJSONString
##  fromList
##  toList
##  print</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb101-1" data-line-number="1"><span class="co"># example of multimedia query using QuerySpec: get the first 100</span></a>
<a class="sourceLine" id="cb101-2" data-line-number="2"><span class="co"># multimedia items associated with a specimen with name starting with "Ba"</span></a>
<a class="sourceLine" id="cb101-3" data-line-number="3">qc &lt;-</a>
<a class="sourceLine" id="cb101-4" data-line-number="4"><span class="st">  </span>QueryCondition<span class="op">$</span><span class="kw">new</span>(<span class="dt">field =</span></a>
<a class="sourceLine" id="cb101-5" data-line-number="5">                       <span class="st">"identifications.scientificName.fullScientificName"</span>,</a>
<a class="sourceLine" id="cb101-6" data-line-number="6">                     <span class="dt">operator =</span></a>
<a class="sourceLine" id="cb101-7" data-line-number="7">                       <span class="st">"STARTS_WITH"</span>,</a>
<a class="sourceLine" id="cb101-8" data-line-number="8">                     <span class="dt">value =</span></a>
<a class="sourceLine" id="cb101-9" data-line-number="9">                       <span class="st">"Qu"</span>)</a>
<a class="sourceLine" id="cb101-10" data-line-number="10">qs &lt;-<span class="st"> </span>QuerySpec<span class="op">$</span><span class="kw">new</span>(<span class="dt">conditions =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(qc), <span class="dt">size =</span> <span class="dv">100</span>)</a>
<a class="sourceLine" id="cb101-11" data-line-number="11">res &lt;-<span class="st"> </span>mc<span class="op">$</span><span class="kw">query</span>(qs)</a>
<a class="sourceLine" id="cb101-12" data-line-number="12"></a>
<a class="sourceLine" id="cb101-13" data-line-number="13"><span class="co"># check if scientific names indeed start with 'Qu'</span></a>
<a class="sourceLine" id="cb101-14" data-line-number="14"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(res<span class="op">$</span>content<span class="op">$</span>resultSet, <span class="cf">function</span>(x)</a>
<a class="sourceLine" id="cb101-15" data-line-number="15">  x<span class="op">$</span>item<span class="op">$</span>identifications[[<span class="dv">1</span>]]<span class="op">$</span>scientificName<span class="op">$</span>fullScientificName)</a></code></pre></div>
<pre><code>##   [1] "Quassia indica (Gaertn.) Noot."                                       
##   [2] "Quassia indica (Gaertn.) Noot."                                       
##   [3] "Quassia indica (Gaertn.) Noot."                                       
##   [4] "Quassia indica (Gaertn.) Noot."                                       
##   [5] "Quassia indica (Gaertn.) Noot."                                       
##   [6] "Quassia borneensis Noot."                                             
##   [7] "Quassia amara L."                                                     
##   [8] "Quassia amara L."                                                     
##   [9] "Quassia baileyana (Oliv.) Noot."                                      
##  [10] "Quassia bidwillii (Hook.f.) Noot."                                    
##  [11] "Quassia amara L."                                                     
##  [12] "Quassia amara L."                                                     
##  [13] "Quassia amara L."                                                     
##  [14] "Quassia indica (Gaertn.) Noot."                                       
##  [15] "Quassia indica (Gaertn.) Noot."                                       
##  [16] "Quassia indica (Gaertn.) Noot."                                       
##  [17] "Quassia indica (Gaertn.) Noot."                                       
##  [18] "Quassia indica (Gaertn.) Noot."                                       
##  [19] "Quassia indica (Gaertn.) Noot."                                       
##  [20] "Quassia indica (Gaertn.) Noot."                                       
##  [21] "Quassia indica (Gaertn.) Noot."                                       
##  [22] "Qualea parviflora Mart."                                              
##  [23] "Qualea polychroma Stafleu"                                            
##  [24] "Quadrimaera pieteri n.sp."                                            
##  [25] "Quadrimaera pieteri n.sp."                                            
##  [26] "Quadrimaera pieteri n.sp."                                            
##  [27] "Quincycolculina distorqucata"                                         
##  [28] "Quelea quelea subsp."                                                 
##  [29] "Quararibea turbinata (Sw.) Poir."                                     
##  [30] "Quercus ilex L."                                                      
##  [31] "Quercus ilex L."                                                      
##  [32] "Quercus ilex L."                                                      
##  [33] "Quercus ilex L."                                                      
##  [34] "Quercus ilex L."                                                      
##  [35] "Quercus ilex L."                                                      
##  [36] "Quercus ilicifolia Wangenh."                                          
##  [37] "Quercus ilicifolia Wangenh."                                          
##  [38] "Quercus imbricaria Michx."                                            
##  [39] "Quercus ithaburensis Decne. subsp. macrolepis (Kotschy) Hedge &amp; Yalt."
##  [40] "Quercus laevis Walter"                                                
##  [41] "Quercus leana Nutt."                                                  
##  [42] "Quercus libani Oliv."                                                 
##  [43] "Quercus libani Oliv."                                                 
##  [44] "Quercus lobata NÃ©e"                                                   
##  [45] "Quercus lusitanica Lam."                                              
##  [46] "Quercus lyrata Walter"                                                
##  [47] "Quercus lyrata Walter"                                                
##  [48] "Quercus macranthera Fisch. &amp; C.A.Mey."                                
##  [49] "Quercus macranthera Fisch. &amp; C.A.Mey."                                
##  [50] "Quercus palustris MÃ¼nchh."                                            
##  [51] "Quercus palustris MÃ¼nchh."                                            
##  [52] "Quercus petraea (Matt.) Liebl."                                       
##  [53] "Quercus petraea (Matt.) Liebl."                                       
##  [54] "Quercus petraea (Matt.) Liebl."                                       
##  [55] "Quercus phellos L."                                                   
##  [56] "Quercus phellos L."                                                   
##  [57] "Quercus phellos L."                                                   
##  [58] "Quercus pontica K.Koch"                                               
##  [59] "Quercus pubescens Willd."                                             
##  [60] "Quercus pubescens Willd."                                             
##  [61] "Quercus pubescens Willd."                                             
##  [62] "Quercus pubescens Willd."                                             
##  [63] "Quercus pubescens Willd."                                             
##  [64] "Quercus pubescens Willd."                                             
##  [65] "Quercus pubescens Willd."                                             
##  [66] "Quercus pubescens Willd."                                             
##  [67] "Quercus pubescens Willd. var. crispata"                               
##  [68] "Quercus pyrenaica Willd."                                             
##  [69] "Quercus pyrenaica Willd."                                             
##  [70] "Quercus pyrenaica Willd."                                             
##  [71] "Quercus robur L."                                                     
##  [72] "Quercus robur L."                                                     
##  [73] "Quercus robur L."                                                     
##  [74] "Quercus rosacea Bechst."                                              
##  [75] "Quercus rosacea Bechst."                                              
##  [76] "Quercus rosacea Bechst."                                              
##  [77] "Quercus suber L."                                                     
##  [78] "Quercus trojana Webb"                                                 
##  [79] "Quercus variabilis Blume"                                             
##  [80] "Quercus velutina Lam."                                                
##  [81] "Quercus velutina Lam."                                                
##  [82] "Quercus velutina Lam."                                                
##  [83] "Quercus velutina Lam."                                                
##  [84] "Quercus velutina Lam."                                                
##  [85] "Quercus velutina Lam."                                                
##  [86] "Quercus velutina Lam."                                                
##  [87] "Quercus libani Oliv."                                                 
##  [88] "Quercus libani Oliv."                                                 
##  [89] "Quercus libani Oliv."                                                 
##  [90] "Quercus macrocarpa Michx."                                            
##  [91] "Quirella  humilis (Preston, 1905)"                                    
##  [92] "Quadrula apiculata (Say, 1829)"                                       
##  [93] "Quassia africana (Baill.) Baill."                                     
##  [94] "Quassia africana (Baill.) Baill."                                     
##  [95] "Quassia africana (Baill.) Baill."                                     
##  [96] "Quassia africana (Baill.) Baill."                                     
##  [97] "Quassia africana (Baill.) Baill."                                     
##  [98] "Quassia africana (Baill.) Baill."                                     
##  [99] "Quassia africana (Baill.) Baill."                                     
## [100] "Quassia africana (Baill.) Baill."</code></pre>
</div>
<div id="data-access-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services-3" class="anchor"></a>Data access services</h3>
<div id="count-3" class="section level4">
<h4 class="hasAnchor">
<a href="#count-3" class="anchor"></a>count</h4>
<p>As for the other data types, a <code>count</code> function returns counts instead of the actual objects:</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb103-1" data-line-number="1"><span class="co"># count all multimedia documents</span></a>
<a class="sourceLine" id="cb103-2" data-line-number="2">mc<span class="op">$</span><span class="kw">count</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] 5722268</code></pre>
</div>
</div>
<div id="aggregation-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services-3" class="anchor"></a>Aggregation services</h3>
<div id="get_distinct_values-3" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values-3" class="anchor"></a>get_distinct_values</h4>
<p>This function returns all values present for a certain field, and their counts. Example: retrieve all different licenses and their counts"</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb105-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_distinct_values</span>(<span class="st">"license"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $`CC0 4.0`
## [1] 5319128
## 
## $`CC BY-NC-ND 4.0`
## [1] 136274
## 
## $`CC BY-NC-SA 4.0`
## [1] 103986
## 
## $`CC BY-NC-SA 3.0`
## [1] 52274
## 
## $`All rights reserved`
## [1] 33973
## 
## $`CC BY-NC-ND 2.5`
## [1] 31592
## 
## $`CC BY-NC-ND`
## [1] 14661
## 
## $`CC BY-SA 4.0`
## [1] 9812
## 
## $`CC BY`
## [1] 9052
## 
## $`CC BY-NC-ND 3.0`
## [1] 4968</code></pre>
</div>
</div>
<div id="metadata-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services-3" class="anchor"></a>Metadata services</h3>
<div id="available-paths-and-fields-2" class="section level4">
<h4 class="hasAnchor">
<a href="#available-paths-and-fields-2" class="anchor"></a>Available paths and fields</h4>
<p>All fields can be retrieved with <code>get_paths</code> and specific information on the fields, such as allowed operators etc. with <code>get_field_info</code></p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb107-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_paths</span>()<span class="op">$</span>content</a>
<a class="sourceLine" id="cb107-2" data-line-number="2">mc<span class="op">$</span><span class="kw">get_field_info</span>()<span class="op">$</span>content</a></code></pre></div>
</div>
<div id="settings-2" class="section level4">
<h4 class="hasAnchor">
<a href="#settings-2" class="anchor"></a>Settings</h4>
<p>Multimedia-specific settings can be retrieved with <code>get_settings</code> and a specific setting with <code>get_setting</code>:</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb108-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_settings</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $index.max_result_window
## [1] 50000</code></pre>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb110-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"index.max_result_window"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
</div>
<div id="operators-2" class="section level4">
<h4 class="hasAnchor">
<a href="#operators-2" class="anchor"></a>Operators</h4>
<p>To test if a certain operator can be used for a mutimedia query:</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb112-1" data-line-number="1">mc<span class="op">$</span><span class="kw">is_operator_allowed</span>(<span class="st">"identifications.scientificName.fullScientificName"</span>,</a>
<a class="sourceLine" id="cb112-2" data-line-number="2">                       <span class="st">"STARTS_WITH"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
</div>
<div id="metadata-services-4" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#metadata-services-4" class="anchor"></a>Metadata services</h2>
<p>Metadata services provide miscellaneous information about the data available via the NBA. Note that there is also type-specific metadata for each data type (e.g. <code>Specimen</code>) which can be retrieved with the specific client of that class. Here we show the available methods for the <code>MetadataClient</code> which gives general, non-type specific metadata. The client is instantiated in the standard way:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb114-1" data-line-number="1">mc &lt;-<span class="st"> </span>MetadataClient<span class="op">$</span><span class="kw">new</span>()</a></code></pre></div>
<div id="controlled-vocabularies" class="section level4">
<h4 class="hasAnchor">
<a href="#controlled-vocabularies" class="anchor"></a>Controlled vocabularies</h4>
<p>The vocabularies for some fields are controlled by dictionaries with allowed values. For the sex of a museum specimen, for instance, only the terms <em>male</em>, <em>female</em>, <em>mixed</em> and <em>hermaphrodite</em> are allowed to be assigned to the specimen. The fields for which controlled lists are available can be retrieved as follows:</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb115-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_lists</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "AreaClass"          "License"            "LicenseType"       
## [4] "RelationType"       "Sex"                "SpatialDatum"      
## [7] "SpecimenTypeStatus" "TaxonomicStatus"</code></pre>
<p>and for each field that has a controlled vocabulary, there is a separate function to retrieve the allowed values:</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb117-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_list_taxonomic_status</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "accepted name"               "synonym"                    
## [3] "basionym"                    "homonym"                    
## [5] "ambiguous synonym"           "misapplied name"            
## [7] "misspelled name"             "provisionally accepted name"</code></pre>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb119-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_list_specimen_type_status</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>##  [1] "allotype"      "epitype"       "hapantotype"   "holotype"     
##  [5] "isoepitype"    "isolectotype"  "isoneotype"    "isosyntype"   
##  [9] "isotype"       "lectotype"     "neotype"       "paratype"     
## [13] "paralectotype" "syntype"       "topotype"      "type"</code></pre>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb121-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_list_sex</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "male"          "female"        "mixed"         "hermaphrodite"
## [5] "unknowable"    "undetermined"</code></pre>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb123-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_controlled_list_phase_or_stage</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## No encoding supplied: defaulting to UTF-8.
## No encoding supplied: defaulting to UTF-8.
## No encoding supplied: defaulting to UTF-8.</code></pre>
<pre><code>## Warning in self$handleError(response): javax.ws.rs.NotFoundException:
## RESTEASY003210: Could not find resource for full path: http://
## api.biodiversitydata.nl/v2/metadata/getControlledList/PhaseOrStage</code></pre>
<pre><code>## [1] "javax.ws.rs.NotFoundException: RESTEASY003210: Could not find resource for full path: http://api.biodiversitydata.nl/v2/metadata/getControlledList/PhaseOrStage"</code></pre>
</div>
<div id="miscellaneous" class="section level4">
<h4 class="hasAnchor">
<a href="#miscellaneous" class="anchor"></a>Miscellaneous</h4>
<div id="dates" class="section level5">
<h5 class="hasAnchor">
<a href="#dates" class="anchor"></a>Dates</h5>
<p>To maintain data integrity, dates have to be coded in specific formats in our systems. For instance, <em>yyyy-MM-dd</em> is a valid format. Allowed formats can be retrieved as follows:</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb127-1" data-line-number="1">mc<span class="op">$</span><span class="kw">get_allowed_date_formats</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [1] "yyyy-MM-dd'T'HH:mm:ssZ"         "yyyy-MM-dd'T'HH:mm[:ss][.SSS]Z"
## [3] "yyyy-MM-dd'T'HH:mm[:ss]"        "yyyy-MM-dd HH:mm[:ss]"         
## [5] "yyyy-MM-dd"                     "yyyy-MM"                       
## [7] "yyyy"</code></pre>
</div>
<div id="services-list" class="section level5">
<h5 class="hasAnchor">
<a href="#services-list" class="anchor"></a>Services list</h5>
<p>The method <code>get_rest_services</code> returns a list of all services in the NBA as objects of type <code>RestService</code>.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb129-1" data-line-number="1"><span class="co"># get the endPoint of the first rest service in the services list</span></a>
<a class="sourceLine" id="cb129-2" data-line-number="2">mc<span class="op">$</span><span class="kw">get_rest_services</span>()<span class="op">$</span>content[[<span class="dv">1</span>]]</a></code></pre></div>
<pre><code>## &lt;RestService&gt;
## Fields:
##  endPoint:    / 
##  method:  GET 
##  consumes:    
##  produces:    text/html;charset=UTF-8 
##  url:     http://api.biodiversitydata.nl/v2/ 
## Methods:
##  fromJSONString
##  toJSONString
##  fromList
##  toList
##  print</code></pre>
</div>
<div id="settings-3" class="section level5">
<h5 class="hasAnchor">
<a href="#settings-3" class="anchor"></a>Settings</h5>
<p>Similar to the document-specific metadata services, we can get general settings with the <code>MetaDataClient</code>:</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb131-1" data-line-number="1"><span class="co">#get all settings</span></a>
<a class="sourceLine" id="cb131-2" data-line-number="2">mc<span class="op">$</span><span class="kw">get_settings</span>()<span class="op">$</span>content</a></code></pre></div>
<pre><code>## $operator.contains.min_term_length
## [1] 3
## 
## $operator.contains.max_term_length
## [1] 15</code></pre>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb133-1" data-line-number="1"><span class="co"># get value for specific setting</span></a>
<a class="sourceLine" id="cb133-2" data-line-number="2">mc<span class="op">$</span><span class="kw">get_setting</span>(<span class="st">"operator.contains.min_term_length"</span>)<span class="op">$</span>content</a></code></pre></div>
<pre><code>## [[1]]
## [1] 3</code></pre>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#services-examples">Services examples</a><ul class="nav nav-pills nav-stacked">
<li><a href="#specimen-occurrence-services">Specimen occurrence services</a></li>
      <li><a href="#taxonomic-data-services">Taxonomic data services</a></li>
      <li><a href="#geographic-data-services">Geographic data services</a></li>
      <li><a href="#multimedia-services">Multimedia services</a></li>
      <li><a href="#metadata-services-4">Metadata services</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hannes Hettling, Rutger Vos.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
